---
title: "R Notebook"
output: html_notebook
---

#Read in HMR data - only use rHMR found in >=50 samples  from Zhao et al #Read in RNAseq and Cibersortx data and merge
```{r}


read_merge_data <- function(log_xform = TRUE){
  
  ####################### Read in and format consensus rHMR data - samples on rows, features on columns - from Aggregate_cpg.R ################################
  # hmr <- read.table(file = "/Users/ds/Desktop/projects/data/bs/all.rHMR.txt", header = T)
  # coordinates <- hmr$Coord
  # hmr <- t(hmr[,-1]) %>% data.frame()
  # colnames(hmr) <- coordinates
  # row.names(hmr) <- row.names(hmr) %>% gsub(pattern = "\\.", replacement = "-")
  # 
  # hmr$IDENTIFIER <- row.names(hmr)


  
  ########################## Read in annotations and genomic data ##############################################
  
  anno1 <- read.csv( "/Users/ds/Desktop/projects/data/anno/Cell2018_supp_T1.csv", header = T) 
  anno2 <- read.csv( "/Users/ds/Desktop/projects/data/anno/Nat_Gen_sampleanno.csv", header = T) 
  anno <- inner_join(anno1, anno2, by = c("IDENTIFIER" = "Sample"))
  anno$Patient.ID <- anno$IDENTIFIER %>% gsub(pattern = "-BL|-PRO|-PRO2", replacement = "")
  
  
  #Read in tumor purity data
  pur <- read.csv("/Users/ds/Desktop/projects/data/anno/WGS_tumor_purity_AF_012318.csv", header = T) %>% dplyr::select(c("WGS.Tumor.Sample.ID","Tumor.Purity.Comp"))
  
  pur$Patient.ID <- pur$WGS.Tumor.Sample.ID %>% gsub(pattern = "-BL|-PRO|-PRO2", replacement = "") 
  
  pur <- pur %>% dplyr::filter(WGS.Tumor.Sample.ID != "DTB-265-BL"	)
  
  
  
  ################### Read in clinical data for OS ##################################
  clin <- read.csv("/Users/ds/Desktop/projects/data/anno/Chen_2019_European_Eurology_Supp4.csv", header = T)
  
  
  
  #################### Read in RNA deconvolution results ###############################

  cf <- cf <- read.table("/Users/ds/Desktop/projects/BME_analysis/results/CIBERSORTx_Job30_Adjusted.txt", sep = "\t", header = T) %>% dplyr::select(-c(P.value, Correlation, RMSE))
  celltypes <- colnames(cf)[-1]
  cf$Mixture <- gsub("\\.", "-", cf$Mixture)
  cf$Patient.ID <- cf$Mixture %>% gsub(pattern = "-BL|-PRO|-PRO2", replacement = "")

  
  
  ###################### Merge data sets based on patient IDs #################################################
  
  #Merge datasets         
  cf <- merge(cf, clin)  %>% merge(pur) %>% merge(anno) #%>% merge(hmr)
  
  #Group/process columns
  cf <- cf %>% mutate(Bone = factor(ifelse(metastasis_biopsy_site == "bone", 1, 0) ))#Format metastatic site
  row.names(cf) <- cf$IDENTIFIER
  
  
  cf <- cf %>% mutate(Enzalutamide.resistant =  as.numeric(Enzalutamide.resistant)) %>% mutate(Type = as.factor(ifelse(Type == "mCRPC", 1, ifelse(Type == "mCRPC (CMP)", 1, 2)))) 
  
  return(cf)
}

cf <- read_merge_data()

celltypes <- c("CD4" , "CD8" , "CTL1" , "CTL2" , "ImmatureB" , "Mature" , "mDC" , "memBcell" , "Mono1" , "Mono2" , "Mono3"  , "Monocyte" , "NK" , "TAM" , "TIM" , "TregR" , "TregA" , "NKT" , "ProB" , "Th117" , "pDC" , "Progenitors",  "Osteoblasts" , "Osteoclasts" ,"Pericytes" , "Endothelial" , "Erythroid", "Tumor" )


```


#Pre-process for DSS
```{r}
set.seed(123)#Set seed for splitting

#Create annotation df for DSS
anno <- cf %>% dplyr::select(c(Bone, Tumor.Purity.Comp, mutation_count, Type, Enzalutamide.resistant, Patient.Age)) 



#Split 70/30
split <- initial_split(anno, prop = 0.7,strata = "Bone" )
anno_train  <- training(split)
anno_test   <- testing(split)



#Write to disk
##anno_train %>% saveRDS("/Users/ds/Desktop/projects/TME_analysis/data/Bone_mCRPC_WGBS_train.rds")

#anno_train %>% cor() %>% cor_plot(label = T)#no evidence of strong correlation among features



```


#Analysis of DSS results
```{r}
#Get output from DSS bed file and convert to GRanges for Methylkit
dmr <- read.table("/Users/ds/Desktop/projects/TME_analysis/results/DMR/all.dmr.Bone.bed", header = T)

dmr <- dmr %>% dplyr::select(c(chr, start,end))
colnames(dmr) <- c("seqnames", "start", "end")

GRanges(dmr) %>% saveRDS("/Users/ds/Desktop/projects/TME_analysis/data/all.dmr.Bone.rds")
```


```{r}
#Read in Methylkit results for DMRs and format
dmr <- read.table("/Users/ds/Desktop/projects/data/bs/all.dmr.Bone.txt", header = T, check.names = F)
coord <- dmr[,1]
dmr <- t(dmr[,-1]) %>% data.frame()
colnames(dmr) <- coord
row.names(dmr) <- gsub("\\.", "-", row.names(dmr))

#Read in dmr to nearest gene df
gene <- read.table(paste0("/Users/ds/Desktop/projects/BME_analysis/results/DMR/all.dmr.Bone.gene.txt"), fill = T)
gene <- gene  %>% mutate(DMR = paste0(V1, ".", V2,".", V3), Genes = V19) %>% dplyr::select(c(DMR, Genes))
  


#Subset training samples for DMR methylation data
dmr_train <- merge(dmr, anno_train, by = 0) 
row.names(dmr_train) <- dmr_train$Row.names
dmr_train <- dmr_train[,-1]
dmr_train <- mutate(dmr_train, Met = ifelse(Bone == 1, "Bone", "Soft_tissue"))
  
#Calculate mean methylation of 2 groups
x <- dmr_train[dmr_train$Bone == 1,] %>% dplyr::select(starts_with("chr")) %>% as.matrix() %>% colMeans()
y <- dmr_train[dmr_train$Bone == 0,]%>% dplyr::select(starts_with("chr")) %>% as.matrix() %>% colMeans()
avg <- data.frame("Bone" = x, "Soft_tissue" = y) 
#Identify DMRs and annotations 
avg <- avg %>% rownames_to_column(var = "DMR") %>% mutate(Diff = Bone - Soft_tissue) %>% mutate(Label = ifelse(Diff < 0 , "Hypo", "Hyper"))


```


#DE-DMR analysis
```{r}
#Get nearest genes to each DMR
avg <- avg %>% merge(gene) 

#Filter DMRs with difference greater than 10%
avg <- avg %>% dplyr::filter(abs(Diff) >=5)


#Get DEG from Limma analysis
query <- dplyr::filter(deg, Sig == "*") %>% .$Genes



#Subset DMRs with nearby genes that are DE
de_dmr <- avg %>% dplyr::filter(Genes %in% query) %>% merge(deg)


#GO ORA of DE-DMRs
query <- de_dmr %>% .$Genes %>% unique()

query


```


```{r}
ego <- enrichGO(gene = query, OrgDb = org.Hs.eg.db, ont = "ALL",keyType = "SYMBOL" ) 

pdf("/Users/ds/Desktop/plot.pdf" , width = 10)
data.frame(ego) %>% arrange(-Count) %>% head(10) %>% ggplot(aes(x = Count, y = reorder(Description, Count), fill = -log10(qvalue))) + geom_col() + scale_fill_viridis_c() + ylab("GO Pathway") + theme(text = element_text(size = 18))
dev.off()

de_dmr

pdf("/Users/ds/Desktop/plot.pdf" , width = 10)
ggplot(de_dmr, aes(x = Diff, y = logFC)) + geom_point() + geom_vline(xintercept = c(-10,10), linetype = "dashed") + geom_hline(yintercept = c(-1,1), linetype = "dashed") + ggrepel::geom_label_repel(data = subset(de_dmr, Genes %in% c( "RUNX2", "COL1A2", "BMP8B")), aes(label = Genes, color = Genes)) + xlab("Methylation Difference") + scale_color_viridis_d(end = 0.5) + theme(axis.title.x = element_text(size = 15), axis.title.y = element_text(size = 15))
dev.off()

pdf("/Users/ds/Desktop/dmr_gene_scrna.pdf" , width = 10)
DoHeatmap(object = scrna, features = unique(de_dmr$Genes), raster = F)
dev.off()



data.frame(ego) %>% arrange(-Count) %>% View()
```





#Running elastic net logistic regression to subset
```{r}

set.seed(123)
X <-  as.matrix(dplyr::select(dmr_train, starts_with("chr")))
y <- as.numeric(ifelse(dmr_train$Bone== 0, -1, 1))


#Logistic regression training
lgr <- cv.glmnet(x =X, y  =dmr_train$Bone, family = "binomial")


lgr_vi <- vi(lgr) %>% head(100) %>% .$Variable


vip(lgr)

```


```{r}

#Get top 20 important features and annotate to nearby genes
lgr_vi <- vi(lgr) %>% dplyr::filter(Importance != 0) %>% mutate(DMR = Variable) %>% merge(gene) %>% arrange(-Importance) %>% mutate(Rank = 1:nrow(.), Method = "LGR")



#Get boxplot of differences in methylation at common important DMRs
dmr_train[,c(lgr_vi$DMR, "Bone")] %>% gather(key = "DMR", value = "Methylation", -Bone) %>% ggplot(aes(x = Bone, y = Methylation, color = Bone)) %>% facet(facet.by = "DMR")+ geom_boxplot() + geom_point(aes(alpha = 0.1), position = position_jitterdodge()) + scale_color_viridis_d(end = 0.5)


pdf("/Users/ds/Desktop/plot.pdf" , width = 10)
DotPlot(object = scrna, features = unique(lgr_vi$Genes), dot.scale = 10)
dev.off()

```


#ROC analysis
```{r}
#ROC/AUC Analysis of DE-DMR

#Subset test samples for DMR methylation data
dmr_test <- merge(dmr, anno_test, by = 0) 
row.names(dmr_test) <- dmr_test$Row.names
dmr_test <- dmr_test[,-1]
dmr_test <- mutate(dmr_test, Met = ifelse(Bone == 1, "Bone", "Soft_tissue"))
```


```{r}


pred_lgr <- predict(lgr,as.matrix(dplyr::select(dmr_test, starts_with("chr"))), type = "response")
rocobj_lgr <- roc(response = dmr_test$Met, predictor = pred_lgr[,1]) 
auc_lgr<- round(auc(dmr_test$Bone, pred_lgr[,1]),4)




#create ROC plot
plot <- ggroc(data = list(LGR = rocobj_lgr)) +
  ggtitle(paste0("\nLR AUC with 5 features = ", auc_lgr)) + 
  geom_abline(slope = 1, intercept = 1, linetype = "dashed") + scale_color_viridis_d(end = 0.5)


ggsave("/Users/ds/Desktop/plot.pdf", plot)




```



```{r}

#Function to calculate AUC of DMRs near genes in GSEA pathways

#Input named list of genes, dmr df for test methylation samples and POI, annotation df of dmr to gene, and avg df of dmr labels
get_auc <- function(gene_list, dmr_df, anno, avg){
  
  
  res <- do.call(rbind, lapply(1:length(gene_list), FUN = function(i){

  gsea_dmr <- anno %>% dplyr::filter(Genes %in% gene_list[[i]]) %>% merge(avg ) %>% .$DMR#Get nearest DMR
  names(gsea_dmr) <- anno %>% dplyr::filter(Genes %in% gene_list[[i]]) %>% merge(avg ) %>% .$Label#Get hypo/hyper label

  #Subset DMRs near select genes  (Bone DMRs identified using training samples in avg df)
 if ( length(gsea_dmr!= 0) ){
    
  #Get z scores for hyper/hypometh regions - flip hypometh so that high Z means bone (1) and low Z means means soft tissue (0)
  hypo <- dmr_df[,c( gsea_dmr[names(gsea_dmr) == "Hypo"])] %>% scale() * -1
  hyper <- dmr_df[,c( gsea_dmr[names(gsea_dmr) == "Hyper"])] %>% scale() 

  
   
  #Get average across all regions as predictions, label is bone/soft tissue
  pred <- cbind(hypo, hyper) %>% rowMeans()
  label <- dmr_df$Bone 

  #Get AUC for label as we sweep across different cutoffs for average of Z-scores for DMRs
  pred <- prediction(pred, label)
  perf <- performance(pred, measure = "auc")
  
  
  if (perf@y.values[[1]] >= 0.5){
      return( c( names(gene_list)[i],  perf@y.values[[1]], paste0(gsea_dmr, collapse = "/"), paste0(gene_list[[i]], collapse = "/") ) )
    
      } else { #Flip Z scores in opposite direction
      
      #Get average across all regions as predictions, label is phenotype of interest
      pred <- cbind(hypo* -1 , hyper* -1) %>% rowMeans()
      label <- dmr_df$Bone
      
      
      #Get AUC for label as we sweep across different cutoffs for average of Z-scores for DMRs
      pred <- prediction(pred, label)
      perf <- performance(pred, measure = "auc")
      return( c( names(gene_list)[i],  perf@y.values[[1]], paste0(gsea_dmr, collapse = "/"), paste0(gene_list[[i]], collapse = "/") ) )
    }
    
  }})) %>% data.frame()
  
 colnames(res) <- c("GO_pathway", "AUC", "DMR", "DEG")
 res$AUC <- as.numeric(res$AUC) 
  
 return(arrange(res, -AUC))
}

```




```{r}
#DMR analysis (unsupervised analysis)
terms <- gsea[grep(gsea$Description, pattern = "bone|collage|ossif"),"ID"]


bone_tme_genes <- gsea %>% dplyr::filter(ID %in% terms) %>% .$core_enrichment %>% str_split("/")#Get terms from GSEA that are related to bone mets
names(bone_tme_genes) <-  gsea %>% dplyr::filter(ID %in% terms) %>% .$Description


res <- get_auc(gene_list = bone_tme_genes, dmr_df = dmr_test, anno = gene, avg = avg)



plot <- res %>% arrange(-AUC) %>% mutate(Rank = 1:nrow(.)) %>% ggplot(aes(x = AUC, y = reorder(GO_pathway, Rank))) + geom_col() + geom_vline(xintercept = 0.5, linetype = "dashed") + ylab("GO pathway")
ggsave("/Users/ds/Desktop/plot.pdf", plot, width = 10)

```
