---
title: "R Notebook"
output: html_notebook
---

```{r}
library(Seurat)
library(ggrepel)
library(msigdbr)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(pROC)
library(edgeR)
library(Signac)
```




```{r}
bmet <- readRDS(file = "/Users/ds/Desktop/scRNA_datasets/He_scRNA/He_mCRPC.rds")
Idents(bmet) <- "organ__ontology_label"


sc_deg <- FindMarkers(object = bmet[row.names(bmet) %in% linc$gene.name,], ident.1 = c("bone tissue", 'lymph node'), ident.2 = 'liver', only.pos = F, verbose = T)
# sc_deg <- sc_deg %>% dplyr::filter(abs(avg_log2FC) > 1 & p_val_adj < 0.1)  %>% mutate(Genes = row.names(.)) %>%  left_join( y = final_markers, by = c("Genes" = "gene")) %>% group_by(cluster.x) %>% arrange(-avg_log2FC.x) %>% slice(1:n())


sc_deg <-  sc_deg %>% dplyr::filter(abs(avg_log2FC) > 1 & p_val_adj < 0.1)  %>% mutate(Genes = row.names(.)) %>%  left_join( y = final_markers, by = c("Genes" = "gene")) %>% arrange(avg_log2FC.x)

#write.csv(x = sc_deg, file = "/Users/ds/Desktop/Supplementary.Table5.csv",quote = F, row.names = F)

Idents(bmet) <- "supercluster.for.LDSC.SEG"

#DotPlot(object = subset(bmet,organ__ontology_label == 'liver' ), features = c("SSTR5-AS1", "LINC00617", "H19", "LINC00514", "FENDRR"))

```

```{r}
mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
mcrpc_rna <- mcrpc_rna[!duplicated(mcrpc_rna$Genename),]




row.names(mcrpc_rna) <- mcrpc_rna$Genename
mcrpc_rna <- mcrpc_rna[,-1]
keep.exprs <- filterByExpr(as.matrix(mcrpc_rna),min.count = 5 )




mcrpc_rna <- mcrpc_rna[keep.exprs,]
mcrpc_rna <- edgeR::cpm(mcrpc_rna, log = T) %>% t() %>% data.frame()#Log normalize for heatmap
#mcrpc_rna <- mcrpc_rna %>% t() %>% data.frame()#Counts for DEG analysis
row.names(mcrpc_rna) <- gsub("\\.", "-",  x = row.names(mcrpc_rna))



anno <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)



anno <- anno %>% dplyr::select(c(sample_id, wgs_id, biopsy_site, disease_type_SIG_GEX, ID_patient,tumor_purity_rna, enza ))
```


```{r}
cf <- read.table("/Users/ds/Desktop/projects/BME_analysis/results/CIBERSORTx_Job33_Adjusted.txt", sep = "\t", header = T)
cf$sample_id <- gsub("\\.", "-", cf$Mixture)
row.names(cf) <- cf$Mixture

cf <- inner_join(x = cf, y = anno, by  = c("sample_id" = "sample_id"))



row.names(cf) <- cf$sample_id




```

```{r}

meta <- read_csv(file = "/Users/ds/Desktop/projects/data/anno/Nat_Gen_sampleanno.csv") %>% data.frame()
row.names(meta) <- meta$Sample


meta <- meta %>% dplyr::select(Type) %>% mutate(Histology = Type) %>% dplyr::select(Histology)
row.names(meta)[meta$Histology == "tSCNC"]
meta
# anno <- cf[!duplicated(cf$wgs_id),] %>% na.omit()
# row.names(anno) <- anno$wgs_id
# anno <- anno %>% dplyr::select(disease_type_SIG_GEX) %>% mutate(Histology = disease_type_SIG_GEX) %>% dplyr::select(Histology)
# 
meta

saveRDS(object = meta, file = "/Users/ds/Desktop/projects/lncRNA/data//Histo_mCRPC_WGBS_anno.rds")

```



```{r}
plot <- ggplot(sc_deg, aes(x = avg_log2FC.x, y = -log10(p_val_adj.x), label = Genes, color = cluster) ) + geom_point(data = dplyr::filter(sc_deg, abs(avg_log2FC.x) > 1 & p_val_adj.x < 0.1), aes(color = cluster)) + scale_color_manual(values = c("purple", "orange", "red", "black", "darkgreen", "navy", "grey")) + geom_hline(yintercept = -log10(0.1), linetype= 'dashed') + geom_vline(xintercept = c(-1,1), linetype = 'dashed') + geom_text_repel(data = dplyr::filter(sc_deg, abs(avg_log2FC.x) > 1 & p_val_adj.x < 0.1 ), force = 0.8, size = 2.5) + xlab("avg_log2FC") + ylab("-log10(p_val_adj)") + ggtitle("lncRNAs associated with Adeno/SC histology") + theme_classic()
 
sc_deg

plot

p <- DotPlot(object = subset(bmet,organ__ontology_label == 'liver' ),features = dplyr::filter(sc_deg,avg_log2FC.x<0 )$Genes ) + coord_flip()

p
liver <- subset(bmet,organ__ontology_label == 'liver' )

sc_sig <- dplyr::filter(sc_deg,avg_log2FC.x<0 )$Genes
adeno_sig <- dplyr::filter(sc_deg,avg_log2FC.x>0 )$Genes


sc_sig
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)
Idents(liver)  <- "supercluster.for.LDSC.SEG"
order <- FindAllMarkers(object = liver, features =sc_sig, verbose = T, only.pos = T)
order <- order%>% group_by(cluster) %>% arrange(-avg_log2FC, .by_group = T)


liver <- ScaleData(object = liver, features =dplyr::filter(sc_deg,avg_log2FC.x<0 )$Genes )
pdf("/Users/ds/Desktop/plot.pdf", width = 10)
DoHeatmap(object =liver,features = c(order$gene[-1], order$gene[1],sc_sig[! sc_sig %in% order$gene]), size = 2.5  )
dev.off()
```







```{r}

adeno <- mcrpc_rna[, colnames(mcrpc_rna) %in% adeno_sig] %>% rowMeans()
sc <- mcrpc_rna[, colnames(mcrpc_rna) %in% sc_sig] %>% rowMeans()
nepc_score <- mcrpc_rna[, colnames(mcrpc_rna) %in% nepc$`HGNC ID`] %>% rowMeans()
shh <- mcrpc_rna[, colnames(mcrpc_rna) %in% dplyr::filter(gs,gs_name == 'HALLMARK_HEDGEHOG_SIGNALING' )$gene_symbol] %>% rowMeans()



df <- data.frame(sample_id = names(adeno), Adeno_signature = adeno, SC_signature = sc, Score = nepc_score, SHH = shh) %>% merge(anno) %>% gather(key = "Signature", value = "GEX", -c(sample_id, biopsy_site, disease_type_SIG_GEX, ID_patient, tumor_purity_rna, wgs_id, Score, SHH)) 


plot <- df%>% ggplot(aes(x = disease_type_SIG_GEX, y = GEX, fill = Signature)) + geom_boxplot() + scale_fill_manual(values = c("navy", "darkred")) + xlab("Histology")

p1 <- dplyr::filter(df, Signature == 'SC_signature')%>% ggplot(aes(x =GEX , y = SHH)) + geom_point() + geom_smooth(method = 'lm') + ggpubr::stat_cor()+ xlab("SC signature") + ylab("SHH MutSigDB pathway")

p2 <- dplyr::filter(df, Signature == 'SC_signature')%>% ggplot(aes(x =GEX , y = Score)) + geom_point() + geom_smooth(method = 'lm') + ggpubr::stat_cor() + xlab("SC signature") + ylab("Beltran NEPC gene set") + theme_classic()



plot <- p1 + p2
p2

ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = p2, width = 10) 

```


```{r}

all_gene_sets = msigdbr(species = "Homo sapiens")

gs <- all_gene_sets %>% dplyr::filter(gs_cat == "H")

```

```{r}
pathways <- gs$gs_name %>% unique()

Idents(bmet) <- "biosample_id"

mat =  AverageExpression(object = bmet)



res <- do.call(rbind, lapply(pathways, FUN = function(p){
  
  print(p)
  g <- gs %>% dplyr::filter(gs_name == p) %>% .$gene_symbol
  df <- mat$RNA %>% gsva( gset.idx.list = list(A = sc_sig, B = adeno_sig, C = g), method = 'zscore') %>% t() %>% data.frame()
  df <- data.frame(Pathway = p, PCC.up = cor.test(df$A,df$C)$estimate, Pval.up = cor.test(df$A,df$C)$p.value, PCC.down = cor.test(df$B,df$C)$estimate, Pval.down = cor.test(df$B, df$C)$p.value)
  return(df)
}))




res %>% arrange(-PCC.up)


res.up <- arrange(res, -PCC.up)[1:10,c("Pathway", "PCC.up")]
colnames(res.up) <- c("Pathway", "PCC")

res.down <- arrange(res, -PCC.down)[1:10,c("Pathway", "PCC.down")]
colnames(res.down) <- c("Pathway", "PCC")

p <- rbind(res.up, mutate(res.down, PCC = PCC * -1)) %>% mutate(Group = ifelse(PCC > 0, "Small cell", "Adeno")) %>% ggplot(aes(y = reorder(Pathway, PCC), x = PCC, fill = Group)) + geom_col() + xlim(c(-1,1)) + scale_x_continuous(breaks = seq(-1,1,0.1),labels = seq(-1,1,0.1) %>% abs()) + ylab("Pathway") + scale_fill_manual(values = c("navy", "darkred")) + theme_classic()
p
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = p, width = 10)


```






```{r}
dss <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/results/all.dmr.Histo.format.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))

dss <- GRanges(dss)



regions <- GRanges(linc)

dss <- data.frame(linc, dss[nearest(x =  GRanges(regions), subject = dss),] ) %>% mutate(Signature = ifelse(gene.name %in% colnames(mcrpc_rna)[colnames(mcrpc_rna) %in% sc_sig], "SC signature", ifelse(gene.name %in% colnames(mcrpc_rna)[colnames(mcrpc_rna) %in% adeno_sig], "Adeno signature", "Other")   ))

dplyr::filter(dss, Signature %in% c("SC signature","Adeno signature")) %>% group_by(Signature) %>% summarise(median(Diff))

plot <- dplyr::filter(dss, Signature %in% c("SC signature","Adeno signature")) %>% ggplot(aes(x = Signature, y= Diff, label = gene.name)) + geom_point() + geom_violin(draw_quantiles = 0.5, aes(fill = Signature, alpha = 0.5)) + geom_jitter(width = 0.1) + xlab("Signature") + ylab("Methylation Difference (SC - Adeno)") +   scale_fill_manual(values = c("navy", "darkred")) + theme_classic()

pdf("/Users/ds/Desktop/plot.pdf", width = 10)
plot
dev.off()
```


```{r}
hmr <- read_table(file = "/Users/ds/Desktop/projects/data/bs/all.mcrpc.hmr.txt")
meta <- meta %>% mutate(wgs_id = row.names(.))

regions <- GRanges(linc)


coord <- hmr$Coord %>% gsub(pattern = "\\.", replacement = "-") %>% StringToGRanges(sep = c("-", "-"))

coord <- data.frame(linc, coord[nearest(x =  GRanges(regions), subject = coord),] ) %>% mutate(Signature = ifelse(gene.name %in% sc_sig, "SC signature", ifelse(gene.name %in% adeno_sig, "Adeno signature", "Other")))
                                                                                                 
                                                                                               


sc_coord <- coord %>% dplyr::filter(Signature == "SC signature") %>% dplyr::select(c(seqnames, start.1, end)) %>% mutate(coord = paste0(seqnames, ".", start.1, ".", end)) %>% .$coord
ad_coord <- coord %>% dplyr::filter(Signature == "Adeno signature") %>% dplyr::select(c(seqnames, start.1, end)) %>% mutate(coord = paste0(seqnames, ".", start.1, ".", end)) %>% .$coord




a <- hmr %>% dplyr::filter(Coord %in% sc_coord) %>% dplyr::select(-Coord) %>% colMeans() %>% data.frame(Dnam = ., wgs_id = gsub("\\.", replacement = "-", names(.))) %>% mutate(Group = "SCNC lncRNAs")
b <- hmr %>% dplyr::filter(Coord %in% ad_coord) %>% dplyr::select(-Coord) %>% colMeans() %>% data.frame(Dnam = ., wgs_id = gsub("\\.", replacement = "-", names(.))) %>% mutate(Group = "Adeno lncRNAs")


p <- rbind(a,b) %>% merge(meta) %>% mutate(Histology = ifelse(Histology == "tSCNC", "SCNC", "Adeno")) %>% ggboxplot(x = "Histology", y = "Dnam", fill = "darkblue", alpha = 0.5, draw_quantiles = 0.5,add = "jitter",width = 0.5, facet.by = "Group") + stat_compare_means(paired = F, method = 't.test', label.x.npc = 0.5) + ylab("Prostate lncRNAs")
nepc



```


```{r}




nepc_lnc <- read_csv(file = "/Users/ds/Desktop/Supplement/beltran_nepc_lnc.csv", col_names = T, skip = 1)
nepc_lnc <- nepc_lnc %>% dplyr::filter(`NEPC (n=122)` == 1)

nepc <- read_csv(file = "/Users/ds/Desktop/Supplement/nepc_genes.csv",col_names = T)




Beltran <- mcrpc_rna[, colnames(mcrpc_rna) %in% nepc$`HGNC ID`] %>% rowMeans()
Ramnarine <- mcrpc_rna[, colnames(mcrpc_rna) %in% unique(nepc_lnc$lncRNA)] %>% rowMeans()
Single_cell <- mcrpc_rna[, colnames(mcrpc_rna) %in% c(adeno_sig,sc_sig)] %>% rowMeans()
SHH <- mcrpc_rna[, colnames(mcrpc_rna) %in% dplyr::filter(gs, gs_name == "HALLMARK_HEDGEHOG_SIGNALING")$gene_symbol] %>% rowMeans()

df <- data.frame(sample_id = names(Beltran), Beltran.et.al = Beltran, Ramnarine.et.al = Ramnarine, Single.cell = Single_cell, Hedgehog = SHH) %>% merge(anno) 


mets <- df  %>% mutate(Histology = disease_type_SIG_GEX) %>% dplyr::select(c(sample_id, Beltran.et.al,Ramnarine.et.al, Single.cell, Histology, Hedgehog )) 


p <- mets %>% dplyr::select(c(Beltran.et.al, Ramnarine.et.al, Single.cell)) %>% gather(key = "Signature" , value = "GEX", -Beltran.et.al ) %>% ggplot(aes(x = GEX, y = Beltran.et.al, color = Signature))  %>% facet(facet.by = "Signature") + geom_point() + geom_smooth(method = 'lm') + stat_cor() + scale_color_manual(values = c("navy", 'darkred')) + theme_classic()

pdf("/Users/ds/Desktop/plot.pdf", width = 10)
p
dev.off()



```

```{r}

nepc_lnc <- read_csv(file = "/Users/ds/Desktop/Supplement/beltran_nepc_lnc.csv", col_names = T, skip = 1)
nepc_lnc <- nepc_lnc %>% dplyr::filter(`NEPC (n=122)` == 1)


nepc <- read_csv(file = "/Users/ds/Desktop/Supplement/nepc_genes.csv",col_names = T)

Beltran =  gsva(t(mcrpc_rna), gset.idx.list = list(c(nepc$`HGNC ID`)), method = 'zscore')
Ramnarine =  gsva(t(mcrpc_rna), gset.idx.list = list(unique(nepc_lnc$lncRNA)), method = 'zscore')
Single_cell = gsva(t(mcrpc_rna), gset.idx.list = list(c(sc_sig, adeno_sig)), method = 'zscore')
SHH = gsva(t(mcrpc_rna), gset.idx.list = list(dplyr::filter(gs, gs_name == "HALLMARK_HEDGEHOG_SIGNALING")$gene_symbol), method = 'zscore')


df <- data.frame(sample_id = colnames(Beltran), Beltran.et.al = t(Beltran), Ramnarine.et.al = t(Ramnarine), Single.cell = t(Single_cell), SHH = t(SHH)) %>% merge(anno) 



mets <- df  %>% mutate(Histology = disease_type_SIG_GEX) %>% dplyr::select(c(sample_id, Beltran.et.al,Ramnarine.et.al, Single.cell, Histology )) 


p <- mets %>% dplyr::select(c(Beltran.et.al, Ramnarine.et.al, Single.cell)) %>% gather(key = "Signature" , value = "GEX", -Beltran.et.al ) %>% ggplot(aes(x = GEX, y = Beltran.et.al, color = Signature))  %>% facet(facet.by = "Signature") + geom_point() + geom_smooth(method = 'lm') + stat_cor() + scale_color_manual(values = c("navy", 'darkred')) + theme_classic()

pdf("/Users/ds/Desktop/plot.pdf", width = 10)
p
dev.off()


```




```{r}

predictions <- mets$Ramnarine.et.al
labels <- mets$Histology

#define object to plot and calculate AUC
rocobj1 <- roc(labels, predictions, direction = "<")
auc1 <- round(auc(labels, predictions, direction = "<"),2)

predictions <- mets$Single.cell

rocobj2 <- roc(labels, predictions, direction = "<")
auc2 <- round(auc(labels, predictions, direction = "<"),2)


predictions <- mets$Beltran.et.al

rocobj3 <- roc(response = labels,predictor =  predictions, direction = ">" )
auc3 <- round(auc(labels, predictions, direction = ">"),2)


#create ROC plot
p <- ggroc(list(Ramnarine.et.al = rocobj1, Single.cell = rocobj2 ), size = 1)+ scale_color_manual(values = c("navy", "darkred", "darkgreen"), name = "Signature", labels = c(paste0("Ramnarine AUC= ", auc1), paste0("Single cell AUC=", auc2), paste0("Beltran AUC=", auc3)) ) + theme_classic() + ggtitle('AUC analysis of 71 CRPC-adeno and 3 CRPC-small cell')
p




```


```{r}
bmet$cluster.dominant.cell.type
Idents(bmet) <- "cluster.dominant.cell.type"
DotPlot(object = bmet, features = c("HOTAIR", "ANRIL", "PCAT19", "MEG3", "TINCR", "PRNCR1", "POLR2E", "HOTTIP", "H19"))
DotPlot(object = bmet, features = c("PCAT19", "PRNCR1"))
DotPlot(object = subset(bmet, organ__ontology_label != 'liver'), features = c("MIAT"))
bmet$organ__ontology_label
final_markers %>% dplyr::filter(gene == "POLR2E")

DotPlot(object = bmet, features = c("H19", "LINC00514", "LINC00617", "SSTR5-AS1"), group.by = "organ__ontology_label")
bmet$organ__ontology_label
mcrpc_rna

DotPlot(object = bmet, features = c("HOTAIR"))

sc_deg
sc_sig[sc_sig == "SSTR5-AS1"]

sc_deg %>% View()
sc_sig
intersect(nepc_lnc$lncRNA, sc_sig)

bmet@meta.data%>% dplyr::select(-biopsy) %>% dplyr::filter(organ__ontology_label == 'liver')

nk.t.markers %>% dplyr::filter(cluster == 'CD8+ PDCD1+ T cell')

sc_sig
nepc_lnc$lncRNA %>% intersect(c(sc_sig, adeno_sig))


DotPlot(object = bmet, features = c("PCA3", "PCAT1", "PCAT29", "PCGEM1", "PRNCR1", "ARLNC1", "ANRIL"), dot.min = .05)




```




