---
title: "R Notebook"
output: html_notebook
---


```{r}


library(TCGAWorkflowData)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(edgeR)
library(EnhancedVolcano)
library(Seurat)
library(lumi)
library(tidyverse)
options(ggrepel.max.overlaps = Inf)

```


```{r}
mrna_query <- GDCquery(project = "TCGA-PRAD",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
mrna_query$results[[1]]


mrna_query$results[[1]] <- mrna_query$results[[1]] %>% dplyr::filter(sample_type == "Solid Tissue Normal")
GDCdownload(mrna_query, method = "api", directory = "/Users/ds/Desktop")

full_path <- "/Users/ds/Desktop/TCGA-PRAD-Normal/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/"



nrml_rna <- do.call(cbind, lapply(list.files(path = full_path), FUN = function(f){
  print(paste0("processing: ",list.files(path = paste0(full_path, f))))
  counts <- read_tsv(file = list.files(path = paste0(full_path, f), full.names = T), skip = 6, col_names = F )
  
  return(data.frame(f = counts[,4]))
  
  
}))


symbols <- read_tsv(file = "/Users/ds/Desktop/TCGA-PRAD-Normal/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/0427d696-a33d-4005-ac1d-c5363754226e/b1becd71-bbed-4a0f-af9c-94b9b2af928b.rna_seq.augmented_star_gene_counts.tsv", skip = 6, col_names = F)[,2]


nrml_rna <- cpm(nrml_rna, log = T, prior.count = 1) %>% data.frame()
colnames(nrml_rna) <- list.files(path = full_path)


nrml_rna <- nrml_rna[which(x =symbols$X2 %in% pca_lncrna ),] 


row.names(nrml_rna) <- symbols$X2[symbols$X2 %in% pca_lncrna]

mrna_query$results[[1]]

nrml_rna <- t(nrml_rna) %>% data.frame()
nrml_rna <- nrml_rna %>% mutate(id = row.names(.), batch = "Normal") %>% merge(dplyr::select(mrna_query$results[[1]] , c(id, cases) ))
                                                               

                                                               
```


#Download TCGA PRAD sample list
```{r}
mrna_query <- GDCquery(project = "TCGA-PRAD",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
mrna_query$results[[1]]


mrna_query$results[[1]] <- mrna_query$results[[1]] %>% dplyr::filter(sample_type == "Primary Tumor")
#GDCdownload(mrna_query, method = "api", directory = "/Users/ds/Desktop")
```


```{r}
full_path <- "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/"

prad_rna <- do.call(cbind, lapply(list.files(path = full_path), FUN = function(f){
  print(paste0("processing: ",list.files(path = paste0(full_path, f))))
  counts <- read_tsv(file = list.files(path = paste0(full_path, f), full.names = T), skip = 6, col_names = F )
  
  return(data.frame(f = counts[,4]))
  
  
}))
#prad_rna <- cpm(prad_rna, log = T, prior.count = 1) %>% data.frame()
colnames(prad_rna) <- list.files(path = full_path)

symbols <- read_tsv(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/0007888f-8d96-4c01-8251-7fef6cc71596/88215dd0-5841-44f1-9393-eefd8238cbb3.rna_seq.augmented_star_gene_counts.tsv", skip = 6, col_names = F)[,2]

prad_rna$Gene <- symbols$X2


#write.table(x = prad_rna, file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_rna_cpm.txt", quote = F, sep = "\t", row.names = F, col.names = T)

write.table(x = prad_rna, file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", quote = F, sep = "\t", row.names = F, col.names = T)
```


#Format PRAD count data
```{r}

#prad_rna <- read.table(file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_rna_cpm.txt", header = T)

prad_rna <- read.table(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", header = T)
ids <- intersect(prad_rna$Gene, linc$gene.name)

prad_rna <- dplyr::filter(prad_rna, Gene %in% ids) #Subset lncRNAs for DEG
prad_rna <- prad_rna[!duplicated(prad_rna$Gene),]


row.names(prad_rna) <- prad_rna$Gene 

prad_rna <- dplyr::select(prad_rna, -Gene)

colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "^X*", replacement = "")

colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "\\.", replacement = "-")



prad_rna <- t(prad_rna) %>% data.frame()

prad_rna <- prad_rna %>% mutate(id = row.names(.), batch = "PRAD") %>% merge(dplyr::select(mrna_query$results[[1]] , c(id, cases.submitter_id) ))
                    
anno_prad <- read.csv(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_supp.csv", header = T)
anno_prad <- anno_prad %>% dplyr::select(avgRNA_purity, PATIENT_ID)


prad_rna <- inner_join(prad_rna, anno_prad, by = c("cases.submitter_id" = "PATIENT_ID")) %>% mutate(Purity = avgRNA_purity * 100, cases = cases.submitter_id)



```




#Format mCRPC count data
```{r}
anno_mcrpc <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)




anno_mcrpc <- anno_mcrpc %>% dplyr::select(c(sample_id, wgs_id, biopsy_site,tumor_purity_wgs, tumor_purity_rna , tumor_purity_hist))


anno_mcrpc <- anno_mcrpc %>% mutate(cases =  sample_id, Purity = tumor_purity_rna)



mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
mcrpc_rna <- mcrpc_rna[!duplicated(mcrpc_rna$Genename),]

row.names(mcrpc_rna) <- mcrpc_rna$Genename
mcrpc_rna <- mcrpc_rna[,-1]
mcrpc_rna <- t(mcrpc_rna) %>% data.frame()
row.names(mcrpc_rna) <- gsub("\\.", "-",  x = row.names(mcrpc_rna))


mcrpc_rna <- mcrpc_rna[,colnames(mcrpc_rna) %in% linc$gene.name,] %>% mutate(cases = row.names(.), batch = "mCRPC")

mcrpc_rna <- mcrpc_rna %>% merge(anno_mcrpc)

query <- intersect(colnames(prad_rna), colnames(mcrpc_rna))



(length(query)-2)/length( linc$gene.name)




rna <- rbind(prad_rna[,query], mcrpc_rna[,query])




```

```{r}
counts <- dplyr::select(rna, -c(cases, Purity, batch)) %>% t() %>% data.frame()
colnames(counts) <- rna$cases



#Create dge object - create counts matrix and sample anno df with library sizes
x <- DGEList(counts = counts)

#Add cluster variable to designate group membership to sample df
x$samples <- cbind(x$samples, dplyr::select(rna, c( Purity, batch )))

#Remove genes with low expression across samples
keep.exprs <- filterByExpr(x)
x <- x[keep.exprs,, keep.lib.size = F]

#Calculate effective library size using EdgeR TMM method
x <- calcNormFactors(x, method = "TMM")

```

```{r}
#Create design matrix (w/o intercept) specifying group membership
design <- model.matrix(~ 0 + batch + Purity, data = rna) 

colnames(design) <- c("mCRPC", "PRAD", "Purity")

# #Create contrasts to calculate
contr.matrix <- makeContrasts(mCRPC - PRAD, levels = colnames(design))


#Normalize counts to log CPM using voom
v <- voom(x, design = design)

#Fit LM according to design
vfit <- lmFit(v, design)

#Calculate contrasts i.e logFC for genes across groups
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

#Calculate t-stats and p vals using Emp bayes method
efit <- eBayes(vfit)


#Get list of all genes, arrange in decreasing order
bulk_deg <- topTable(efit, number = Inf, coef = 1) %>% data.frame(Genes = row.names(.))  %>% arrange(-logFC) 
bulk_deg <- bulk_deg %>% arrange(-logFC)



```


```{r}
up <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC > 1) %>% .$Genes
down <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC < -1)  %>% arrange(logFC) %>% .$Genes



plot <- EnhancedVolcano(toptable = bulk_deg, x = "logFC", y = "adj.P.Val", lab =row.names(bulk_deg), selectLab = c(up[1:20], down[1:20]), pointSize = 1 , max.overlaps	= Inf, drawConnectors = T, pCutoff = 1e-2, legendPosition = "none", title = "Differential expression analysis of lncRNAs during progression", subtitle = "mCRPC v.s PRAD", FCcutoff = 1, labSize = 3) + scale_color_viridis_d(option = "plasma", end = 0.8)
plot
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)
```


```{r}
#Map up and down regulated genes to clusters
up <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC > 1) %>% left_join( y = filt_markers, by = c("Genes" = "gene"))
down <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC < -1)  %>% arrange(logFC) %>% left_join( y = filt_markers, by = c("Genes" = "gene"))



f <- up$Genes                                          
names(f) <- up$cluster
f <- f[!is.na(names(f))]


f <- f[!duplicated(f)]
p1 <- DotPlot(object = bmet, features =  f)

f <- down$Genes                                          
names(f) <- down$cluster
f <- f[!is.na(names(f))]


f <- f[!duplicated(f)]
p2 <- DotPlot(object = bmet, features =  f)


plot <- plot_grid(p1, p2, ncol = 1, byrow = T, labels = list("Upregulated", "Downregulated"))

plot

p1

pdf("/Users/ds/Desktop/plot.pdf", height = 10, width = 15)
p1  + theme(axis.text.x = element_text(size = 8, angle = 45, vjust = 0.75))  + ggtitle("Upregulated in mCRPC")
dev.off()




```


```{r}


dss <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/results/DSS_mCPRC_PRAD.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))
dss
dss <- GRanges(dss)
regions <- GRanges(linc)


regions <- resize(x = regions, width = width(regions) + 5000, fix = "center")

dss <- data.frame(linc, dss[nearest(x =  GRanges(regions), subject = dss),] ) %>% inner_join(bulk_deg, by = c("gene.name" = "Genes" )) %>% mutate(Sig = ifelse(gene.name %in% up$Genes & Diff < -0.05, "A", ifelse(gene.name %in% down$Genes & Diff > 0.05, "B", "Other")   ))





plot <- dss %>% ggplot(aes(x = Diff * 100, y = logFC, color = Sig, label= gene.name, alpha = Sig)) + geom_point(size= 1) + geom_vline(xintercept = c(-5,5), linetype = "dashed")  + geom_hline(yintercept = c(-1,1), linetype = "dashed") + geom_text_repel(data = dplyr::filter(dss, gene.name %in% c("MALAT1", "H19", "DNM3OS", "LUCAT1", "PVT1", "FTX")),  color = "black", force =10, nudge_y = 10) + scale_alpha_manual(values = c(1,1,0.1))  + ylab("logFC mCRPC v.s PRAD") + xlab("Methylation difference mCRPC v.s PRAD") +  scale_color_viridis_d(option = "plasma", end = 0.5)
plot

View(dss)
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)

dss %>% dplyr::filter(Sig == "A") %>% View()

up
```


```{r}
mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/bulkRNA_mCRPC_CPM_gene.txt", header = T)

# 
# all_gene_sets = msigdbr(species = "Homo sapiens")
# 
# gs <- all_gene_sets %>% dplyr::filter(gs_cat == "H")
# 
# querygenes <- list(TGFB = gs[grepl(pattern = "TGF", x = gs$gs_name),"gene_symbol", drop = T])
# 
# 
# a <- mcrpc_rna %>% dplyr::filter(Gene %in% querygenes$TGFB) %>% dplyr::select(-Gene) %>% log1p()  %>% colMeans()


b <- mcrpc_rna %>% dplyr::filter(Gene %in% "FTX") %>% dplyr::select(-Gene) %>% log1p() %>% colMeans()



df <- data.frame(FTX = b, sample_id = gsub(pattern = "\\.", replacement = "-", x = names(b)))  %>% merge(cf) 
#df <- mutate(df, RB1 = ifelse(RB1 == 0, "Biallelic inactivation", "Monoallelic inactivation/wildtype"))
df

bmet$cluster.dominant.cell.type
DotPlot(object = bmet, features = "FTX", group.by = "cluster.dominant.cell.type")
p1 <- df %>% ggplot(aes(x = FTX, color = CD4 + CD8 + TregA + TregR + CTL1 + CTL2 + NK + NKT +  Th117, y =  TGFB )) + geom_point() + geom_smooth(method = "lm", color = "black") + stat_cor() + scale_color_viridis_c(option = 'plasma', name = "Fraction NK and T cell") + ylab("MutSigDB TGFB GEX") + xlab("HCP5 GEX")


stat.test <- df %>% wilcox_test(formula = TGFB ~ RB1) %>% add_significance()   
p2 <- ggboxplot(df, x = "RB1", y = "TGFB", fill = viridis::cividis(10)[5])
stat.test <- stat.test %>% add_xy_position(x = "RB1")
p2 <- p2 + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))


stat.test <- df %>% wilcox_test(formula = HCP5 ~ RB1) %>% add_significance()   
p3 <- ggboxplot(df, x = "RB1", y = "HCP5", fill = viridis::plasma(10)[5])
stat.test <- stat.test %>% add_xy_position(x = "RB1")
p3 <- p3 + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))

plot <- p1 + p2  + p3


```




