---
title: "R Notebook"
output: html_notebook
---


```{r}


library(TCGAWorkflowData)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(edgeR)
library(EnhancedVolcano)
library(Seurat)
library(lumi)
library(tidyverse)
options(ggrepel.max.overlaps = Inf)

```


```{r}
mrna_query <- GDCquery(project = "TCGA-PRAD",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
mrna_query$results[[1]]


mrna_query$results[[1]] <- mrna_query$results[[1]] %>% dplyr::filter(sample_type == "Solid Tissue Normal")
GDCdownload(mrna_query, method = "api", directory = "/Users/ds/Desktop")

full_path <- "/Users/ds/Desktop/TCGA-PRAD-Normal/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/"



nrml_rna <- do.call(cbind, lapply(list.files(path = full_path), FUN = function(f){
  print(paste0("processing: ",list.files(path = paste0(full_path, f))))
  counts <- read_tsv(file = list.files(path = paste0(full_path, f), full.names = T), skip = 6, col_names = F )
  
  return(data.frame(f = counts[,4]))
  
  
}))


symbols <- read_tsv(file = "/Users/ds/Desktop/TCGA-PRAD-Normal/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/0427d696-a33d-4005-ac1d-c5363754226e/b1becd71-bbed-4a0f-af9c-94b9b2af928b.rna_seq.augmented_star_gene_counts.tsv", skip = 6, col_names = F)[,2]


nrml_rna <- cpm(nrml_rna, log = T, prior.count = 1) %>% data.frame()
colnames(nrml_rna) <- list.files(path = full_path)


nrml_rna <- nrml_rna[which(x =symbols$X2 %in% pca_lncrna ),] 


row.names(nrml_rna) <- symbols$X2[symbols$X2 %in% pca_lncrna]

mrna_query$results[[1]]

nrml_rna <- t(nrml_rna) %>% data.frame()
nrml_rna <- nrml_rna %>% mutate(id = row.names(.), batch = "Normal") %>% merge(dplyr::select(mrna_query$results[[1]] , c(id, cases) ))
                                                               

                                                               
```



```{r}
mrna_query <- GDCquery(project = "TCGA-PRAD",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
mrna_query$results[[1]]


mrna_query$results[[1]] <- mrna_query$results[[1]] %>% dplyr::filter(sample_type == "Primary Tumor")
#GDCdownload(mrna_query, method = "api", directory = "/Users/ds/Desktop")
```


```{r}
full_path <- "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/"

prad_rna <- do.call(cbind, lapply(list.files(path = full_path), FUN = function(f){
  print(paste0("processing: ",list.files(path = paste0(full_path, f))))
  counts <- read_tsv(file = list.files(path = paste0(full_path, f), full.names = T), skip = 6, col_names = F )
  
  return(data.frame(f = counts[,4]))
  
  
}))
#prad_rna <- cpm(prad_rna, log = T, prior.count = 1) %>% data.frame()
colnames(prad_rna) <- list.files(path = full_path)

symbols <- read_tsv(file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/0007888f-8d96-4c01-8251-7fef6cc71596/88215dd0-5841-44f1-9393-eefd8238cbb3.rna_seq.augmented_star_gene_counts.tsv", skip = 6, col_names = F)[,2]

prad_rna$Gene <- symbols$X2


#write.table(x = prad_rna, file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_rna_cpm.txt", quote = F, sep = "\t", row.names = F, col.names = T)

write.table(x = prad_rna, file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", quote = F, sep = "\t", row.names = F, col.names = T)
```


```{r}

#prad_rna <- read.table(file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_rna_cpm.txt", header = T)

prad_rna <- read.table(file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", header = T)
ids <- intersect(prad_rna$Gene, linc$gene.name)

prad_rna <- dplyr::filter(prad_rna, Gene %in% ids) #Subset lncRNAs for DEG
prad_rna <- prad_rna[!duplicated(prad_rna$Gene),]


row.names(prad_rna) <- prad_rna$Gene 

prad_rna <- dplyr::select(prad_rna, -Gene)

colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "^X*", replacement = "")

colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "\\.", replacement = "-")



prad_rna <- t(prad_rna) %>% data.frame()

prad_rna <- prad_rna %>% mutate(id = row.names(.), batch = "PRAD") %>% merge(dplyr::select(mrna_query$results[[1]] , c(id, cases.submitter_id) ))
                    
anno_prad <- read.csv(file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_supp.csv", header = T)
anno_prad <- anno_prad %>% dplyr::select(avgRNA_purity, PATIENT_ID)


prad_rna <- inner_join(prad_rna, anno_prad, by = c("cases.submitter_id" = "PATIENT_ID")) %>% mutate(Purity = avgRNA_purity * 100, cases = cases.submitter_id)


```





```{r}
anno_mcrpc <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)




anno_mcrpc <- anno_mcrpc %>% dplyr::select(c(sample_id, wgs_id, biopsy_site,tumor_purity_wgs, tumor_purity_rna , tumor_purity_hist, disease_type_SIG_GEX, ID_patient, enza, AR_enhancer, TMPRSS2_ERG))


anno_mcrpc <- anno_mcrpc %>% mutate(cases =  sample_id, Purity = tumor_purity_rna)



mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
mcrpc_rna <- mcrpc_rna[!duplicated(mcrpc_rna$Genename),]

row.names(mcrpc_rna) <- mcrpc_rna$Genename
mcrpc_rna <- mcrpc_rna[,-1]
mcrpc_rna <- t(mcrpc_rna) %>% data.frame()
row.names(mcrpc_rna) <- gsub("\\.", "-",  x = row.names(mcrpc_rna))


mcrpc_rna <- mcrpc_rna[,colnames(mcrpc_rna) %in% linc$gene.name,] %>% mutate(cases = row.names(.), batch = "mCRPC")

mcrpc_rna <- mcrpc_rna %>% merge(anno_mcrpc)

query <- intersect(colnames(prad_rna), colnames(mcrpc_rna))



(length(query)-2)/length( linc$gene.name)




rna <- rbind(prad_rna[,query], mcrpc_rna[,query])




```

```{r}
counts <- dplyr::select(rna, -c(cases, Purity, batch)) %>% t() %>% data.frame()
colnames(counts) <- rna$cases



#Create dge object - create counts matrix and sample anno df with library sizes
x <- DGEList(counts = counts)

#Add cluster variable to designate group membership to sample df
x$samples <- cbind(x$samples, dplyr::select(rna, c( Purity, batch )))

#Remove genes with low expression across samples
keep.exprs <- filterByExpr(x)
x <- x[keep.exprs,, keep.lib.size = F]

#Calculate effective library size using EdgeR TMM method
x <- calcNormFactors(x, method = "TMM")

```

```{r}
#Create design matrix (w/o intercept) specifying group membership
design <- model.matrix(~ 0 + batch + Purity, data = rna) #+ biopsy_site, data = cf) #+ disease_type_SIG_GEX, data = cf)

design
colnames(design) <- c("mCRPC", "PRAD", "Purity")

# #Create contrasts to calculate
contr.matrix <- makeContrasts(mCRPC - PRAD, levels = colnames(design))


#Normalize counts to log CPM using voom
v <- voom(x, design = design)

#Fit LM according to design
vfit <- lmFit(v, design)

#Calculate contrasts i.e logFC for genes across groups
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

#Calculate t-stats and p vals using Emp bayes method
efit <- eBayes(vfit)


#Get list of all genes, arrange in decreasing order
bulk_deg <- topTable(efit, number = Inf, coef = 1) %>% data.frame(Genes = row.names(.))  %>% arrange(-logFC) 
bulk_deg <- bulk_deg %>% arrange(-logFC)



```


```{r}
up <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC > 1) %>% .$Genes
down <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC < -1)  %>% arrange(logFC) %>% .$Genes



plot <- EnhancedVolcano(toptable = bulk_deg, x = "logFC", y = "adj.P.Val", lab =row.names(bulk_deg), selectLab = c(up[1:20], down[1:20]), pointSize = 1 , max.overlaps	= Inf, drawConnectors = T, pCutoff = 1e-2, legendPosition = "none", title = "Differential expression analysis of lncRNAs during progression", subtitle = "mCRPC v.s PRAD", FCcutoff = 1, labSize = 3)
plot
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)

up
down

pdf("/Users/ds/Desktop/plot.pdf", height = 10)
DotPlot(object = bmet, features = up[1:30], assay = "RNA") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + coord_flip()
dev.off()
```



```{r}


bulk_deg %>% dplyr::filter(Genes %in% c(up, down))  %>% mutate(Group = ifelse(Genes %in% up, "Up", "Down"))  %>% left_join(y = filt_markers, by = c("Genes" = "gene"))%>% group_by(Group, cluster) %>% mutate(Count = n()) %>%  ggplot( aes(fill=cluster, y=Count, x=Group)) + geom_bar(position="dodge", stat="identity")

```


```{r}
hmr_gene <- data.frame(linc, hmr[nearest(x =  GRanges(linc), subject = hmr),] )
hmr_gene <- hmr_gene %>% mutate(Coord =paste0(seqnames, ".", start.1, ".", end) )
?contrasts.fit
hmr_meth %>% dplyr::filter(Coord %in% hmr_gene$Coord) %>% View()


dss <- read_tsv(file = "/Users/ds/Desktop/DNAm_datasets/DSS_mCPRC_PRAD.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))

dss <- GRanges(dss)



dss <- data.frame(linc, dss[nearest(x =  GRanges(linc), subject = dss),] ) %>% inner_join(bulk_deg, by = c("gene.name" = "Genes" )) %>% mutate(Sig = ifelse(gene.name %in% c(up) & Diff < -0.05, "A", ifelse(gene.name %in% c(down) & Diff > 0.05, "B", "Other")   ))


plot <- dss %>% ggplot(aes(x = Diff * 100, y = logFC, color = Sig, label= gene.name, alpha = Sig)) + geom_point(size= 1) + geom_vline(xintercept = c(-5,5), linetype = "dashed")  + geom_hline(yintercept = c(-1,1), linetype = "dashed") + geom_text_repel(data = dplyr::filter(dss, gene.name %in% c("MALAT1", "H19", "DNM3OS", "LUCAT1", "NEAT1", "SCHLAP1", "HAGLR", "MEG3", "PCAT14")), color = "black", force =200) + scale_alpha_manual(values = c(1,1,0.5)) + scale_color_manual(values = c("darkblue", "darkred", "darkgrey")) + ylab("logFC mCRPC v.s PRAD") + xlab("Methylation difference mCRPC v.s PRAD")
plot
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)

dss %>% dplyr::filter(Sig == "B") %>% View()


```




```{r}
mcrpc_meth <- read_tsv(file = "/Users/ds/Desktop/projects/data/bs/all.mcrpc.hmr.txt", col_names = T)
mcrpc_meth <- mcrpc_meth[!duplicated(mcrpc_meth$Coord),]


prad_meth <- read_tsv(file = "/Users/ds/Desktop/projects/data/bs/all.prad.hmr.txt", col_names = T)



#mcrpc_meth <- mcrpc_meth #%>% group_by(Coord) #%>% mutate(Coord = paste0(Coord, ":", 1:n())) %>% ungroup()




#prad_meth <- prad_meth #%>% group_by(Coord) #%>% mutate(Coord = paste0(Coord, ":", 1:n())) %>% ungroup()


hmr_meth <- inner_join(mcrpc_meth, prad_meth, by = c("Coord" = "Coord"))


mval <- hmr_meth %>% dplyr::select(starts_with(c("DTB", "SR")))

mval <- beta2m(mval/100)

row.names(mval) <- hmr_meth$Coord
colnames(mval) <- colnames(hmr_meth %>% dplyr::select(starts_with(c("DTB", "SR"))))

anno <- colnames(mval) %>% data.frame(ID = ., Group = ifelse(grepl(pattern = "DTB", x = colnames(mval)), "mCRPC", "PRAD"))

```


```{r}


hmr_meth

#Create design matrix (w/o intercept) specifying group membership
design <- model.matrix(~ Group, data = anno) 


vfit <- lmFit(mval, design)

#Calculate contrasts i.e logFC for genes across groups
#vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

#Calculate t-stats and p vals using Emp bayes method
efit <- eBayes(vfit)

#Get list of all genes, arrange in decreasing order
dmr <- topTable(efit, number = Inf, coef = 2) %>% data.frame(Coord = row.names(.))  %>% arrange(-logFC)
dmr
ncol(design)


View(dmr)
dmr <- dmr %>% inner_join(hmr_gene, by = c("Coord" = "Coord"))
dmr
hmr_gene
intersect(dmr$Coord, hmr_gene$Coord)

coord <- dmr %>% dplyr::filter(gene.name %in% c(up,down))  %>% .$Coord

df <- hmr_meth %>% dplyr::filter(Coord %in% coord) %>% rowwise() %>% mutate( mCRPCvsPRAD =  mean(c_across(contains("DTB."))) - mean(c_across(contains("SRR")))) %>% dplyr::select(c(Coord, mCRPCvsPRAD)) %>% merge(dmr) %>% mutate(Group = ifelse(gene.name %in% up, "Up in mCRPC", "Up in PRAD"), Sig = ifelse(adj.P.Val < 0.01, "*", "ns")) 

bulk_deg

df %>% ggplot(aes( x= mCRPCvsPRAD, y = -log10(adj.P.Val) , alpha = Sig, label = gene.name, color = Group)) + geom_point(size = 1) + scale_alpha_discrete(range=c(1, 0.01)) + geom_text_repel(data = dplyr::filter(df, Sig == "*" & abs(mCRPCvsPRAD) > 20) ) + scale_color_manual(values= c("darkblue", "darkred"))


dplyr::filter(df, Group == "Up in mCRPC" & mCRPCvsPRAD < -5) %>% View()
View(data.frame(hmr))


hmr_meth %>% dplyr::filter(Coord == "chr14.100850800.100851500") %>% gather(key = "ID", value = "Meth" , -Coord) %>% mutate(Batch = ifelse(grepl(pattern = "DTB", x = ID), "mCRPC", "PRAD")) %>% ggplot(aes(x  = Batch, y = Meth)) + geom_boxplot()


df %>% dplyr::filter(gene.name == "MALAT1")
```




```{r}
ehmr <- read.csv(file = "/Users/ds/Desktop/DNAm_datasets/ehmr.csv", header = T)%>% mutate(Coord = paste0(chrom, ".", start, ".",end ))

query <- ehmr %>% dplyr::filter(symbol %in% c(up, down) & type == "HMR")  %>% .$Coord


View(ehmr)
query

mat <- dplyr::select(hmr_meth, starts_with(c("DTB", "SR"))) %>% t() %>% data.frame()
colnames(mat) <- hmr_meth$Coord




mat <- mat[,grepl(pattern = paste0(query, collapse = "|"), x = colnames(mat))] %>% mutate(ID = row.names(.)) %>% mutate(Batch = ifelse(grepl(pattern= "DTB", x = ID), "mCRPC", "PRAD") )



mat <- mat %>% gather(key = "Coord", value = "DNAm", -c(ID, Batch))



res <- do.call(rbind, lapply(unique(mat$Coord), FUN= function(r){
  

  df <- dplyr::filter(mat, Coord == r)
  
  #Fit linear regression model with covariate as bone met site 
  model <- lm(formula = df$DNAm ~ df$Batch) 
  #Get coeff for Bone and pvalue
  coef <- summary(model)$coefficients[, "Estimate"][2]
  pval <- summary(model)$coefficients[, "Pr(>|t|)"][2]
  
  return(  data.frame(Coord  = r, Coef = coef, Pval = pval))

 })) %>% data.frame()



#FDR correction across celltypes
res$Adj.Pvalue <-p.adjust(res$Pval, method = "BH")
res$Sig <-  ifelse(res$Adj.Pvalue < 0.1, "*", "ns")




res <- res %>% mutate(Coord = gsub(pattern = ":1", replacement = "", x = Coord)) %>% merge(ehmr)  %>% mutate(Group = ifelse(symbol %in% up, "Up in mCRPC", "Up in PRAD")) %>% group_by(symbol) %>% mutate(Label = paste0(symbol, ":", 1:n())) %>% ungroup()


up



```


```{r}
plot <- ggplot(res, aes(x = Coef, y = -log10(Pval), color = Group, alpha = Sig, label = symbol)) + geom_point() + scale_color_manual(values = c("darkred", "darkblue")) + xlab("Methylation difference in PRAD - mCRPC") + geom_text_repel(data = dplyr::filter(res, Sig == "*" & symbol %in% c(down, up)& abs(Coef) > 10) ) + scale_alpha_discrete(range=c(1, 0.1))

plot
res %>% View()
plot
```

