---
title: "R Notebook"
output: html_notebook
---


```{r}
counts <- dplyr::select(rna, -c(cases, Purity, batch)) %>% t() %>% data.frame()
colnames(counts) <- rna$cases



#Create dge object - create counts matrix and sample anno df with library sizes
x <- DGEList(counts = counts)

#Add cluster variable to designate group membership to sample df
x$samples <- cbind(x$samples, dplyr::select(rna, c( Purity, batch )))

#Remove genes with low expression across samples
keep.exprs <- filterByExpr(x)
x <- x[keep.exprs,, keep.lib.size = F]

#Calculate effective library size using EdgeR TMM method
x <- calcNormFactors(x, method = "TMM")
x <- x[row.names(x) %in% linc$gene.name,]#Subset lncrna in hg38 ###############



#Create design matrix (w/o intercept) specifying group membership
design <- model.matrix(~ 0 + batch + Purity, data = rna) 

colnames(design) <- c("mCRPC", "PRAD", "Purity")

# #Create contrasts to calculate
contr.matrix <- makeContrasts(mCRPC - PRAD, levels = colnames(design))


#Normalize counts to log CPM using voom
v <- voom(x, design = design)

#Fit LM according to design
vfit <- lmFit(v, design)

#Calculate contrasts i.e logFC for genes across groups
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

#Calculate t-stats and p vals using Emp bayes method
efit <- eBayes(vfit)


#Get list of all genes, arrange in decreasing order
bulk_deg <- topTable(efit, number = Inf, coef = 1) %>% data.frame(Genes = row.names(.))  %>% arrange(-logFC) 
bulk_deg <- bulk_deg %>% arrange(-logFC)



```


```{r}
up <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC > 1) %>% .$Genes
down <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC < -1)  %>% arrange(logFC) %>% .$Genes



plot <- EnhancedVolcano(toptable = bulk_deg, x = "logFC", y = "adj.P.Val", lab =row.names(bulk_deg), selectLab = c(up[1:20], down[1:20]), pointSize = 1 , max.overlaps	= Inf, drawConnectors = T, pCutoff = 1e-2, legendPosition = "none", title = "Differential expression analysis of lncRNAs during progression", subtitle = "mCRPC v.s PRAD", FCcutoff = 1, labSize = 3) 
plot
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10, height = 10)
```


```{r}
#Map up and down regulated genes to clusters
up <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC > 1) %>% left_join( y = final_markers, by = c("Genes" = "gene"))
down <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.1 & logFC < -1)  %>% arrange(logFC) %>% left_join( y = final_markers, by = c("Genes" = "gene"))



na.omit(down)
```


#Map to cell types
```{r}
up_ae <- AverageExpression(object = bmet, features = na.omit(up)[,"Genes"], slot = "data") %>%  data.frame() %>% log1p() 


up_ae <- up_ae %>% data.frame() %>% mutate(Genes = row.names(.)) %>% inner_join(up, by = c("Genes" = "Genes")) %>% dplyr::select(c(starts_with("RNA"), "Genes", "cluster")) 
up_ae <- mutate(up_ae, Group = "Upregulated")

down_ae <- AverageExpression(object = bmet, features = na.omit(down)[,"Genes"], slot = "data") %>%  data.frame() %>% log1p() 


down_ae <- down_ae %>% data.frame() %>% mutate(Genes = row.names(.)) %>% inner_join(down, by = c("Genes" = "Genes")) %>% dplyr::select(c(starts_with("RNA"), "Genes", "cluster")) 

down_ae <- mutate(down_ae, Group = "Downregulated")

ae <- rbind(up_ae, down_ae)
ae <- ae[!duplicated(ae$Genes),]


ae <- ae %>% group_by(Group) %>% arrange(cluster, .by_group = T) %>% ungroup() %>% data.frame()
row.names(ae) <- ae$Genes





pheatmap(mat = dplyr::select(ae, starts_with("RNA")), annotation_row = dplyr::select(ae, c(cluster, Group)), cluster_cols = F, cluster_rows = F, scale = "row", gaps_row = nrow(dplyr::filter(ae, Group == "Downregulated")), filename = "/Users/ds/Desktop/plot.png", height = 10)


```

```{r}
f <- up$Genes                                          
names(f) <- up$cluster
f <- f[!is.na(names(f))]


f <- f[!duplicated(f)]
p1 <- DotPlot(object = bmet, features =  f)

f <- down$Genes                                          
names(f) <- down$cluster
f <- f[!is.na(names(f))]


f <- f[!duplicated(f)]
p2 <- DotPlot(object = bmet, features =  f)


plot <- plot_grid(p1, p2, ncol = 1, byrow = T, labels = list("Upregulated", "Downregulated"))

plot

p1

pdf("/Users/ds/Desktop/plot.pdf", height = 10, width = 15)
p1  + theme(axis.text.x = element_text(size = 8, angle = 45, vjust = 0.75))  + ggtitle("Upregulated in mCRPC")
dev.off()




```


```{r}


dss <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/results/DSS_mCPRC_PRAD.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))
dss
dss <- GRanges(dss)
regions <- GRanges(linc)


regions <- resize(x = regions, width = width(regions) + 5000, fix = "center")

dss <- data.frame(linc, dss[nearest(x =  GRanges(regions), subject = dss),] ) %>% inner_join(bulk_deg, by = c("gene.name" = "Genes" )) %>% mutate(Sig = ifelse(gene.name %in% up$Genes & Diff < -0.05, "A", ifelse(gene.name %in% down$Genes & Diff > 0.05, "B", "Other")   ))





plot <- dss %>% ggplot(aes(x = Diff * 100, y = logFC, color = Sig, label= gene.name, alpha = Sig)) + geom_point(size= 1) + geom_vline(xintercept = c(-5,5), linetype = "dashed")  + geom_hline(yintercept = c(-1,1), linetype = "dashed") + geom_text_repel(data = dplyr::filter(dss, gene.name %in% c("MALAT1", "H19", "DNM3OS", "LUCAT1", "PVT1", "FTX")),  color = "black", force =10, nudge_y = 10) + scale_alpha_manual(values = c(1,1,0.1))  + ylab("logFC mCRPC v.s PRAD") + xlab("Methylation difference mCRPC v.s PRAD") +  scale_color_viridis_d(option = "plasma", end = 0.5)
plot

View(dss)
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)

dss %>% dplyr::filter(Sig == "A") %>% View()

up
```


```{r}
mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/bulkRNA_mCRPC_CPM_gene.txt", header = T)

# 
# all_gene_sets = msigdbr(species = "Homo sapiens")
# 
# gs <- all_gene_sets %>% dplyr::filter(gs_cat == "H")
# 
# querygenes <- list(TGFB = gs[grepl(pattern = "TGF", x = gs$gs_name),"gene_symbol", drop = T])
# 
# 
# a <- mcrpc_rna %>% dplyr::filter(Gene %in% querygenes$TGFB) %>% dplyr::select(-Gene) %>% log1p()  %>% colMeans()


b <- mcrpc_rna %>% dplyr::filter(Gene %in% "FTX") %>% dplyr::select(-Gene) %>% log1p() %>% colMeans()



df <- data.frame(FTX = b, sample_id = gsub(pattern = "\\.", replacement = "-", x = names(b)))  %>% merge(cf) 
#df <- mutate(df, RB1 = ifelse(RB1 == 0, "Biallelic inactivation", "Monoallelic inactivation/wildtype"))
df

bmet$cluster.dominant.cell.type
DotPlot(object = bmet, features = "FTX", group.by = "cluster.dominant.cell.type")
p1 <- df %>% ggplot(aes(x = FTX, color = CD4 + CD8 + TregA + TregR + CTL1 + CTL2 + NK + NKT +  Th117, y =  TGFB )) + geom_point() + geom_smooth(method = "lm", color = "black") + stat_cor()  + ylab("MutSigDB TGFB GEX") + xlab("HCP5 GEX")


stat.test <- df %>% wilcox_test(formula = TGFB ~ RB1) %>% add_significance()   
p2 <- ggboxplot(df, x = "RB1", y = "TGFB", fill = viridis::cividis(10)[5])
stat.test <- stat.test %>% add_xy_position(x = "RB1")
p2 <- p2 + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))


stat.test <- df %>% wilcox_test(formula = HCP5 ~ RB1) %>% add_significance()   
p3 <- ggboxplot(df, x = "RB1", y = "HCP5", fill = viridis::plasma(10)[5])
stat.test <- stat.test %>% add_xy_position(x = "RB1")
p3 <- p3 + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))

plot <- p1 + p2  + p3


```





