---
title: "R Notebook"
output: html_notebook
---

```{r}
mcrpc_cpg <- read.table("/Users/ds/Desktop/projects/data/bs/chr11.mcrpc.cpg.txt", header = T)
prad_cpg <- read.table("/Users/ds/Desktop/projects/data/bs/chr11.prad.cpg.txt", header = T)



query <- intersect(mcrpc_cpg$Coord, prad_cpg$Coord) %>% str_split("\\.") %>% unlist() %>% matrix(ncol = 3, byrow = T) %>% data.frame() 



colnames(query) <- c("Chr", "Start", "End")
query$Chr <- gsub(pattern = "chr", replacement = "", x = query$Chr)

```

```{r}
GOI <- "FTX"
chr <- "chrX"

target <- EnsDb.Hsapiens.v86 %>% ensembldb::filter(~ symbol == GOI) %>% genes()
#target <- resize(x = target, width = width(target) + 10000, fix = "center")
target <- target[1]
```


```{r}


dss <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/results/DSS_mCPRC_PRAD.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))
dss <- dss %>% dplyr::filter(chrom == chr) %>% dplyr::select(-Stat) %>% mutate(Diff = Diff * -100) %>% GRanges()


dTrack <- DataTrack(dss, name = "DSS", background.title = viridis::plasma(10)[8], fill = viridis::plasma(10)[8])

plotTracks(list(dTrack), from = start(target), to = end(target), type = c("histogram"), baseline = 0)

```


```{r}


overlaps <-subsetByOverlaps(x = GRanges(query), ranges = target) 



overlaps <- overlaps%>% data.frame() %>% mutate(Coord = paste0("chr", seqnames, ".", start, ".", end))
row.names(overlaps) <- overlaps$Coord


regions <- cbind(dplyr::filter(prad_cpg, Coord %in% overlaps$Coord), dplyr::select(dplyr::filter(mcrpc_cpg, Coord %in% overlaps$Coord), -Coord))



row.names(regions) <- regions$Coord
regions <- regions[,-1]

colnames(regions) <- gsub(pattern= "\\.", replacement = "-", x = colnames(regions))
regions <- t(regions) %>% data.frame()





coord <- regions %>% t() %>% row.names() %>% StringToGRanges(sep = c("\\.", "\\."))



coord <- GRanges(coord, metadata = t(regions))




dTrack <- DataTrack(coord, name = "CpG", background.title = "DarkOrchid4")

plotTracks(list(dTrack), from = start(target), to = end(target), groups = factor(ifelse( grepl(pattern = "SR", x = row.names(regions)), "PRAD", "mCRPC")), type = c("smooth"), legend=TRUE, lwd = 2)


```







```{r}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene %>% exons(columns=c("GENEID", "EXONID", "TXNAME" ))



txdb <- S4Vectors::expand(txdb, c("GENEID", "EXONID", "TXNAME"))
# seqlevelsStyle(target) <- "UCSC"
# gene <- subsetByOverlaps(x = txdb, ranges = target) %>% data.frame()
```


```{r}
gene <- txdb[txdb$GENEID ==  100302692] %>% data.frame()
colnames(gene) <- c("chromosome", "start", "end", "width", "strand", "gene", "exon", "transcript")



hmr <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/data/hmr.bed", col_names  = c("chrom", "start", "end"))
hmr <- hmr %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()

metarbs <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/data/AR_chip_hg38.bed", col_names = c("chrom", "start", "end"))
metarbs <- metarbs %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()

atac <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/data/ATAC_mCRPC_consensus.bed", col_names = c("chrom", "start", "end"))
atac <- atac %>% dplyr::filter(chrom == chr) %>% dplyr::select(c(chrom, start, end)) %>% GRanges()


```



```{r}


#Plot
grtrack <- GeneRegionTrack( genome = "hg38",range = gene, chromosome = chr , name = GOI, transcriptAnnotation = "transcript", background.panel = "lightyellow", background.title = viridis::plasma(10)[1], fill = viridis::plasma(10)[1], stacking = "dense", col = viridis::plasma(10)[1])



atrack <- AnnotationTrack(hmr, name = "HMR", background.title = viridis::plasma(10)[3],  background.panel ="lightgrey", fill =viridis::plasma(10)[3], col = viridis::plasma(10)[3])

arbtrack <- AnnotationTrack(metarbs, name = "ARBS", background.title = viridis::plasma(10)[6] , background.panel = "lightgrey", fill = viridis::plasma(10)[6], col = viridis::plasma(10)[6])
atactrack <- AnnotationTrack(atac, name = "ATAC", background.title = viridis::plasma(10)[4] , background.panel = "lightgrey", fill = viridis::plasma(10)[4], rotation = 90, col = viridis::plasma(10)[4])

gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome = "hg38", chromosome = chr)



pdf("/Users/ds/Desktop/plot.pdf", width = 10)
plotTracks(list(itrack,gtrack, dTrack,arbtrack, atactrack, atrack, grtrack), from = start(target)- 1000, to = end(target) + 1000, type = "histogram")
dev.off()   
           


```



