---
title: "R Notebook"
output: html_notebook
---

```{r}
library(rstatix)
library(ggpubr)
```




```{r}
cf <- read.table("/Users/ds/Desktop/projects/BME_analysis/results/CIBERSORTx_Job33_Adjusted.txt", sep = "\t", header = T)
cf$sample_id <- gsub("\\.", "-", cf$Mixture)
row.names(cf) <- cf$Mixture

anno <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)



anno <- anno %>% dplyr::select(c(sample_id, wgs_id, biopsy_site, disease_type_SIG_GEX, ID_patient, RB1, tumor_purity_rna, tumor_purity_wgs, tumor_purity_hist)) 

cf <- inner_join(x = cf, y = anno, by  = c("sample_id" = "sample_id"))

row.names(cf) <- cf$sample_id


cf <- na.omit(cf)

cf <- cf %>% mutate(RB1 = ifelse(RB1 == 0, 0, 1))

cf

```

#Subset WGBS samples for DMR analysis
```{r}
anno <- cf[!duplicated(cf$wgs_id),]
row.names(anno) <- anno$wgs_id
anno <- anno %>% dplyr::select(RB1) %>% mutate(RB1 = as.factor(RB1))

saveRDS(object = anno, file = "/Users/ds/Desktop/projects/lncRNA/data//RB1_mCRPC_WGBS_anno.rds")

```


```{r}


mcrpc_rna <- mcrpc_rna %>% dplyr::filter(Gene %in% c(up,down)) 
genes <- mcrpc_rna$Gene

mcrpc_rna <- mcrpc_rna %>% dplyr::select(-Gene) %>% t() %>% data.frame() 
colnames(mcrpc_rna) <- genes


mcrpc_rna <- mcrpc_rna %>% mutate(sample_id = gsub(pattern = "\\.", replacement = "-", x = row.names(.))) %>% merge(cf)  %>% na.omit()

mcrpc_rna
res <- do.call(rbind, lapply(genes, FUN= function(g){
  
  print(g)
  #Fit linear regression model with covariate as bone met site 
  model <- lm(formula = log1p(mcrpc_rna[,g]) ~ mcrpc_rna[,"enza"]) 
  #Get coeff for Bone and pvalue
  coef <- summary(model)$coefficients[, "Estimate"][2]
  pval <- summary(model)$coefficients[, "Pr(>|t|)"][2]
  
  return(  data.frame(Gene  = g, Coef = coef, Pval = pval))

 })) %>% data.frame()



#FDR correction across celltypes
res$Adj.Pvalue <-p.adjust(res$Pval, method = "BH")
res$Sig <-  ifelse(res$Adj.Pvalue < 0.1, "*", "ns")


#%>% na.omit() %>% ggplot(aes(x = enza, y = Mono1 + Mono2 + Mono3 + TAM + TIM + mDC + Monocyte + CD4 + CD8 + TregA + TregR + CTL1 + CTL2 + NK + NKT +  Th117)) + geom_boxplot() 

#%>% ggplot(aes(x = log1p(RNA), y = Mono1 + Mono2 + Mono3 + TAM + TIM + mDC + Monocyte + CD4 + CD8 + TregA + TregR + CTL1 + CTL2 + NK + NKT +  Th117, color = enza  )) + geom_point() + stat_cor() + geom_smooth(method = "lm")

res

```




```{r}
#Limma DE analysis 


#Read in raw counts, convert to matrix with rows as gene symbols, columns as samples
counts <- read.table("/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)


counts <- counts[!duplicated(counts$Genename),]
row.names(counts) <- counts$Genename
counts <- counts[,-1] %>% as.matrix()
colnames(counts) <- gsub("\\.", "-", colnames(counts))  

counts <- counts[row.names(counts) %in% linc$gene.name,]#Subset lncrna in hg38 

#Keep matching sample ids of RNA-seq and WGS for cmp annotation
ids <- intersect(colnames(counts),row.names(cf))
counts <- counts[,ids]
cf <- cf[ids,]


#Create dge object - create counts matrix and sample anno df with library sizes
x <- DGEList(counts = counts)

#Add cluster variable to designate group membership to sample df
x$samples <- cbind(x$samples, dplyr::select(cf, c( RB1 )))

#Remove genes with low expression across samples
keep.exprs <- filterByExpr(x)
x <- x[keep.exprs,, keep.lib.size = F]

#Calculate effective library size using EdgeR TMM method
x <- calcNormFactors(x, method = "TMM")

#Create design matrix (w/o intercept) specifying group membership
design <- model.matrix(~ 0 + as.factor(RB1) ,data = cf) #+ biopsy_site, data = cf) #+ disease_type_SIG_GEX, data = cf)

design
colnames(design) <- c("Del", "WT") #, "Liver", "LN", "Other")

# #Create contrasts to calculate
contr.matrix <- makeContrasts(Del - WT, levels = colnames(design))

#Normalize counts to log CPM using voom
v <- voom(x, design = design)

#Fit LM according to design
vfit <- lmFit(v, design)

#Calculate contrasts i.e logFC for genes across groups
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

#Calculate t-stats and p vals using Emp bayes method
efit <- eBayes(vfit)


#Get list of all genes, arrange in decreasing order
bulk_deg <- topTable(efit, number = Inf, coef = 1) %>% data.frame(Genes = row.names(.))  %>% arrange(-logFC) 
bulk_deg <- bulk_deg %>% arrange(-logFC)

#bulk_deg %>% dplyr::filter(adj.P.Val < 0.25 & abs(logFC) > 1) %>%   left_join(y = filt_markers, by = c("Genes" = "gene"))
up <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.25 & logFC > 1) %>% left_join( y = filt_markers, by = c("Genes" = "gene"))
down <- bulk_deg %>% dplyr::filter(adj.P.Val < 0.25 & logFC < -1)  %>% arrange(logFC) %>% left_join( y = filt_markers, by = c("Genes" = "gene"))
```




```{r}


f <- up$Genes                                          
names(f) <- up$cluster
f <- f[!is.na(names(f))]


f <- f[!duplicated(f)]
p1 <- DotPlot(object = bmet, features =  f)+ theme(axis.text.x = element_text(size = 8))  
p1
f <- down$Genes                                          
names(f) <- down$cluster
f <- f[!is.na(names(f))]


f <- f[!duplicated(f)]
p2 <- DotPlot(object = bmet, features =  f)+ theme(axis.text.x = element_text(size = 8))  

plot <- plot_grid(p1, p2, ncol = 1, byrow = T, labels = list("Upregulated", "Downregulated"))


pdf("/Users/ds/Desktop/plot.pdf", height = 15, width = 15)
plot 
dev.off()
```



```{r}
plot <-  EnhancedVolcano(toptable = bulk_deg, x = "logFC", y = "adj.P.Val", lab =row.names(bulk_deg), selectLab = c( up$Genes[1:10], down$Genes[1:10]), pointSize = 1 , max.overlaps	= Inf, drawConnectors = T, pCutoff = 25e-02, legendPosition = "none", title = "Differential expression analysis of lncRNAs and RB1 mutations", subtitle = "RB1 biallelic inactivation v.s monoallelic/wildtype mCRPCs", FCcutoff = 1, labSize = 3)+ scale_color_viridis_d(option = "plasma")
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 20)

cf
up

plot
DotPlot(object = bmet, features = "CDKN2B-AS1")
up
```



```{r}
mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/bulkRNA_mCRPC_CPM_gene.txt", header = T)


all_gene_sets = msigdbr(species = "Homo sapiens")

gs <- all_gene_sets %>% dplyr::filter(gs_cat == "H")

querygenes <- list(TGFB = gs[grepl(pattern = "TGF", x = gs$gs_name),"gene_symbol", drop = T])


a <- mcrpc_rna %>% dplyr::filter(Gene %in% querygenes$TGFB) %>% dplyr::select(-Gene) %>% log1p()  %>% colMeans()
b <- mcrpc_rna %>% dplyr::filter(Gene %in% "HCP5") %>% dplyr::select(-Gene) %>% log1p() %>% colMeans()



df <- data.frame(TGFB = a,HCP5 = b, sample_id = gsub(pattern = "\\.", replacement = "-", x = names(b)))  %>% merge(cf) 
df <- mutate(df, RB1 = ifelse(RB1 == 0, "Biallelic inactivation", "Monoallelic inactivation/wildtype"))


p1 <- df %>% ggplot(aes(x = HCP5, color = CD4 + CD8 + TregA + TregR + CTL1 + CTL2 + NK + NKT +  Th117, y =  TGFB )) + geom_point() + geom_smooth(method = "lm", color = "black") + stat_cor() + scale_color_viridis_c(option = 'plasma', name = "Fraction NK and T cell") + ylab("MutSigDB TGFB GEX") + xlab("HCP5 GEX")


stat.test <- df %>% wilcox_test(formula = TGFB ~ RB1) %>% add_significance()   
p2 <- ggboxplot(df, x = "RB1", y = "TGFB", fill = viridis::cividis(10)[5])
stat.test <- stat.test %>% add_xy_position(x = "RB1")
p2 <- p2 + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))


stat.test <- df %>% wilcox_test(formula = HCP5 ~ RB1) %>% add_significance()   
p3 <- ggboxplot(df, x = "RB1", y = "HCP5", fill = viridis::plasma(10)[5])
stat.test <- stat.test %>% add_xy_position(x = "RB1")
p3 <- p3 + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))

plot <- p1 + p2  + p3


```

```{r}



dss <- read_tsv(file = "/Users/ds/Desktop/projects/lncRNA/results/all.dmr.RB1.format.bed", col_names = c("chrom", "start", "end", "Diff", "Stat"))

dss <- GRanges(dss)
regions <- GRanges(linc)


regions <- resize(x = regions, width = width(regions) + 5000, fix = "center")



dss <- data.frame(linc, dss[nearest(x =  GRanges(regions), subject = dss),] ) %>% inner_join(bulk_deg, by = c("gene.name" = "Genes" )) %>% mutate(Sig = ifelse(gene.name %in% up$Genes & Diff < -0.05, "A", ifelse(gene.name %in% down$Genes & Diff > 0.05, "B", "Other")   ))



plot <- dss %>% ggplot(aes(x = Diff * 100, y = logFC, color = Sig, label= gene.name, alpha = Sig)) + geom_point(size= 1) + geom_vline(xintercept = c(-5,5), linetype = "dashed")  + geom_hline(yintercept = c(-1,1), linetype = "dashed") + geom_text_repel(data = dplyr::filter(dss, gene.name %in% c("SCHLAP1", "PCAT1", "PCAT14", "CCAT2", "PRNCR1", "PART1", "ARLNC1", "PCAT4", "HOTAIR")), color = "black", force =10) + scale_alpha_manual(values = c(1,1,0.1))  + ylab("logFC RB1 loss v.s wildtype") + xlab("Methylation difference RB1 loss - wildtype") + scale_color_viridis_d(option = "plasma", end = 0.5)
plot

dss %>% dplyr::filter(Sig == "A")
ggsave(filename = "/Users/ds/Desktop/plot.pdf", plot = plot, width = 10)

```



