---
title: "R Notebook"
output: html_notebook
---

#Load libraries
```{r}


library(TCGAWorkflowData)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(edgeR)
library(EnhancedVolcano)
library(Seurat)
library(lumi)
library(tidyverse)
library(ggrepel)
options(ggrepel.max.overlaps = Inf)

```


#Load TCGA annotations
```{r}
mrna_query <- GDCquery(project = "TCGA-PRAD",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
mrna_query$results[[1]]



nrml_meta <- mrna_query$results[[1]] %>% dplyr::filter(sample_type == "Solid Tissue Normal")
#GDCdownload(mrna_query, method = "api", directory = "/Users/ds/Desktop")
```

#Process normal prostate counts
```{r}
full_path <- "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Normal/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/"



nrml_rna <- do.call(cbind, lapply(list.files(path = full_path), FUN = function(f){
  print(paste0("processing: ",list.files(path = paste0(full_path, f))))
  counts <- read_tsv(file = list.files(path = paste0(full_path, f), full.names = T), skip = 6, col_names = F )
  
  return(data.frame(f = counts[,4]))
  
  
}))


symbols <- read_tsv(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Normal/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/0427d696-a33d-4005-ac1d-c5363754226e/b1becd71-bbed-4a0f-af9c-94b9b2af928b.rna_seq.augmented_star_gene_counts.tsv", skip = 6, col_names = F)[,2]
colnames(nrml_rna) <- list.files(path = full_path)




nrml_rna$Gene <- symbols$X2

write.table(x = nrml_rna, file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Normal/normal_rna_counts.txt", quote = F, sep = "\t", row.names = F, col.names = T)
```

#Change IDS for normals
```{r}
nrml_rna <- read.table(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Normal/normal_rna_counts.txt", header = T)


nrml_rna <- nrml_rna[!duplicated(nrml_rna$Gene),]


row.names(nrml_rna) <- nrml_rna$Gene 

nrml_rna <- dplyr::select(nrml_rna, -Gene)

colnames(nrml_rna) <- colnames(nrml_rna) %>% gsub(pattern = "^X*", replacement = "")

colnames(nrml_rna) <- colnames(nrml_rna) %>% gsub(pattern = "\\.", replacement = "-")



nrml_rna <-  edgeR::cpm(nrml_rna, log = T) %>% t() %>% data.frame()#Log normalize
#nrml_rna <- nrml_rna %>% t() %>% data.frame()

nrml_rna <- nrml_rna %>% mutate(id = row.names(.), batch = "Normal") %>% merge(dplyr::select(nrml_meta , c(id, cases.submitter_id) ))
nrml_rna <- nrml_rna %>% mutate(cases = cases.submitter_id, Purity = NA, Grade_Mutational_status =NA)
```



#Download TCGA PRAD sample list
```{r}

prad_meta<- mrna_query$results[[1]] %>% dplyr::filter(sample_type == "Primary Tumor")
#GDCdownload(mrna_query, method = "api", directory = "/Users/ds/Desktop")
```



#Process PRAD counts
```{r}
full_path <- "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/"

prad_rna <- do.call(cbind, lapply(list.files(path = full_path), FUN = function(f){
  print(paste0("processing: ",list.files(path = paste0(full_path, f))))
  counts <- read_tsv(file = list.files(path = paste0(full_path, f), full.names = T), skip = 6, col_names = F )
  
  return(data.frame(f = counts[,4]))
  
  
}))
#prad_rna <- cpm(prad_rna, log = T, prior.count = 1) %>% data.frame()
colnames(prad_rna) <- list.files(path = full_path)

symbols <- read_tsv(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification/0007888f-8d96-4c01-8251-7fef6cc71596/88215dd0-5841-44f1-9393-eefd8238cbb3.rna_seq.augmented_star_gene_counts.tsv", skip = 6, col_names = F)[,2]

prad_rna$Gene <- symbols$X2


#write.table(x = prad_rna, file = "/Users/ds/Desktop/TCGA/TCGA-PRAD-Tumor/prad_rna_cpm.txt", quote = F, sep = "\t", row.names = F, col.names = T)

write.table(x = prad_rna, file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", quote = F, sep = "\t", row.names = F, col.names = T)
```


#Format PRAD count data
```{r}

prad_rna <- read.table(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_rna_counts.txt", header = T)
prad_rna <- prad_rna[!duplicated(prad_rna$Gene),]


row.names(prad_rna) <- prad_rna$Gene 

prad_rna <- dplyr::select(prad_rna, -Gene)

colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "^X*", replacement = "")

colnames(prad_rna) <- colnames(prad_rna) %>% gsub(pattern = "\\.", replacement = "-")

prad_rna <- edgeR::cpm(prad_rna, log = T) %>% t() %>% data.frame()#Log normalize for heatmap
#prad_rna <- prad_rna %>% t() %>% data.frame()#Counts for DEG analysis



prad_rna <- prad_rna %>% mutate(id = row.names(.), batch = "PRAD") %>% merge(dplyr::select(prad_meta , c(id, cases.submitter_id) ))




```



#Add PRAD purity from GDC website
```{r}
anno_prad <- read.csv(file = "/Users/ds/Desktop/projects/TCGA/TCGA-PRAD-Tumor/prad_supp.csv", header = T)
anno_prad <- anno_prad %>% dplyr::select(avgRNA_purity, PATIENT_ID, Reviewed_Gleason_sum)

prad_rna <- inner_join(prad_rna, anno_prad, by = c("cases.submitter_id" = "PATIENT_ID")) %>% mutate(Purity = avgRNA_purity * 100, cases = cases.submitter_id, Grade_Mutational_status = Reviewed_Gleason_sum)


```



#Format mCRPC count data
```{r}
anno_mcrpc <- read.table("/Users/ds/Desktop/projects/data/anno/202009_deepRNAseq_sample_full.txt", header =T)




anno_mcrpc <- anno_mcrpc %>% dplyr::select(c(sample_id, wgs_id, biopsy_site,tumor_purity_wgs, tumor_purity_rna , tumor_purity_hist, AR))


anno_mcrpc <- anno_mcrpc %>% mutate(cases =  sample_id, Purity = tumor_purity_rna, Grade_Mutational_status =AR) %>% dplyr::select(cases, Purity, Grade_Mutational_status)



mcrpc_rna <- read.table(file = "/Users/ds/Desktop/projects/data/rna/mCRPC_RNA_counts_genename.txt", header = T)
mcrpc_rna <- mcrpc_rna[!duplicated(mcrpc_rna$Genename),]




row.names(mcrpc_rna) <- mcrpc_rna$Genename
mcrpc_rna <- mcrpc_rna[,-1]
mcrpc_rna <- edgeR::cpm(mcrpc_rna, log = T) %>% t() %>% data.frame()#Log normalize for heatmap
#mcrpc_rna <- mcrpc_rna %>% t() %>% data.frame()#Counts for DEG analysis
row.names(mcrpc_rna) <- gsub("\\.", "-",  x = row.names(mcrpc_rna))



mcrpc_rna <- mcrpc_rna %>% mutate(cases = row.names(.), batch = "mCRPC")

#Add mCRPC purity
mcrpc_rna <- mcrpc_rna %>% merge(anno_mcrpc)



```

#Merge PRAD and mCRPC datasets (DEG analysis)
```{r}

query <- intersect(colnames(prad_rna), colnames(mcrpc_rna))#Get lncRNAs in common + cases + purity

(length(query)-2)/length( linc$gene.name) #% of lncRNAs common across datasets

#Combine PRAD and mCRPC - rows are samples and columns are lncRNAs + RNA purity
rna <- rbind(prad_rna[,query], mcrpc_rna[,query])


prad_rna$Grade_Mutational_status
```

#Merge PRAD, Normal, and mCRPC datasets (Heatmap)
```{r}
query <- intersect(colnames(prad_rna), colnames(mcrpc_rna)) %>% intersect( colnames(nrml_rna))#Get lncRNAs in common + cases + purity


#Combine PRAD and mCRPC - rows are samples and columns are lncRNAs + RNA purity + batch
comb_rna <- rbind( nrml_rna[,query], arrange(prad_rna[,query],Grade_Mutational_status), arrange(mcrpc_rna[,query],Grade_Mutational_status))


#comb_rna <- data.frame(comb_rna[,c("batch", "Purity", "cases", "Grade_Mutational_status")], comb_rna[,colnames(comb_rna) %in% c(pca_lncrna, notpca_lncrna)])

comb_rna <- data.frame(comb_rna[,c("batch", "Purity", "cases", "Grade_Mutational_status")], comb_rna[,colnames(comb_rna) %in% c(pca_lncrna, notpca_lncrna, "AR", "KLK3", "TMPRSS2", "FOXA1")])

comb_rna <- comb_rna %>% mutate(cases = paste0(cases, ":", 1:nrow(comb_rna)))


row.names(comb_rna) <- comb_rna$cases



```




#Plot lncRNAs in TME
```{r}

anno_col <- dplyr::select(comb_rna, c(batch, Purity, Grade_Mutational_status))


colnames(anno_col) <- c("Batch", "RNA.Tumor.Purity", "GS.Mutational.status")
anno_col <- mutate(anno_col, GS.Mutational.status = ifelse(GS.Mutational.status == 2, "AR WT", ifelse(GS.Mutational.status == 3, "AR Gain", GS.Mutational.status)))



norm_ids <- comb_rna %>% dplyr::filter(batch == "Normal") %>% .$cases %>% gsub(pattern = "-|:",replacement = "\\.")



mat <- comb_rna[,colnames(comb_rna) %in% c(up$Genes, down$Genes)] %>% t() %>% data.frame() %>% mutate(gene = row.names(.)) %>% inner_join(final_markers) %>% arrange(cluster) 
row.names(mat) <-mat$gene 



anno_row <- mat %>% dplyr::select(cluster)
colnames(anno_row) <- c("Cell.type")

mat <- mat %>% dplyr::select(starts_with(c("TCGA", "DTB", "PR")))


norm_meds <- mat[norm_ids] %>%  as.matrix() %>% rowMedians(useNames = T)



mat <- apply(mat, 2, FUN = function(x){x - norm_meds  })



# filt <- mat[apply(mat, 1, FUN = function(x){sum(x < 0)}) == ncol(mat),] %>% row.names()
# filt
# mat <- mat[!rownames(mat) %in% filt,]
# anno_row <- anno_row[!rownames(anno_row) %in% filt,, drop = F]
# 

column_ha = HeatmapAnnotation( RNA.Tumor.Purity = anno_col$RNA.Tumor.Purity, Batch = anno_col$Batch, GS.Mutational.status = anno_col$GS.Mutational.status)
row_ha <- rowAnnotation(Cell.type = anno_row$Cell.type)


pdf("/Users/ds/Desktop/plot.pdf", width = 10, height = 15)
Heatmap(matrix = mat, show_column_names = F, col  = colorRampPalette(c('navy',"white", "red"))(10), top_annotation = column_ha, cluster_columns = F,left_annotation = row_ha, row_names_gp = gpar(fontsize = 10,col = ifelse(rownames(mat) %in% c(unique(up$Genes)) , "red", ifelse(rownames(mat)  %in% c(unique(down$Genes)), "blue", "black"))), row_split = anno_row$Cell.type, name = "logFC")
dev.off()


```



```{r}

bmet$cluster.dominant.cell.type
nk.t <- subset(bmet, supercluster.for.LDSC.SEG == 'NK/T')
Idents(nk.t) <- "cluster.dominant.cell.type"

nk.t.markers <- FindAllMarkers(object = nk.t[row.names(nk.t) %in% linc$gene.name,], only.pos = T, verbose = T)

nk.t.markers  %>% dplyr::filter(pct.1 > 0.1 & p_val_adj < 0.1 & avg_log2FC> 1)  %>% dplyr::filter(cluster == 'CD8+ PDCD1+ T cell')

p1 <- DotPlot(object = nk.t, features = c( 'AL135999.1', 'PVT1', 'MIAT', 'LINC00861' , 'AC008105.3', 'AL109811.2', 'FTX', 'AC116366.1', 'MIATNB', 'AC020916.1', 'CYTOR', 'LINC00954', 'BISPR', 'AC026979.2'), group.by ="cluster.dominant.cell.type" ) + coord_flip() + theme(axis.text.x = element_text(size = 8, angle = 90))


exhaust <- mcrpc_rna[, colnames(mcrpc_rna) %in% c("PDCD1", "HAVCR2", "TOX", "TIGIT", "ICOS", "FASLG", "LAG3", "ENTPD1","ITGAE"), drop = F] %>% rowMeans()      

p2 <- data.frame(GEX=rowMeans(mcrpc_rna[,c("MIAT", "CYTOR"), drop = F] ),Exhausted.markers=exhaust ) %>% ggplot(aes(x = GEX, y = Exhausted.markers)) + geom_point() + geom_smooth(method = 'lm') + stat_cor() + xlab("CYTOR/MIAT") + theme_classic()

p2



pdf("/Users/ds/Desktop/plot.pdf", width = 15)
p1 + p2
dev.off()

```


```{r}
df1 <- comb_rna %>% group_by(batch) %>% arrange(AR, .by_group = T)  %>% dplyr::filter(batch == "Normal")

df2 <- comb_rna %>% group_by(batch) %>% arrange(AR, .by_group = T)  %>% dplyr::filter(batch == "PRAD")
df3 <- comb_rna %>% group_by(batch) %>% arrange(AR, .by_group = T)  %>% dplyr::filter(batch == "mCRPC")
```


```{r}
df <- rbind(df1, df2, df3) %>% ungroup() %>% mutate(Index = 1:n()) 


#lm(AR ~ Purity, data = dplyr::filter(comb_rna, batch != 'Normal')) %>% residuals() %>% data.frame(AR = .)

#df <- dplyr::filter(df, batch != 'Normal') 


df <- df%>% dplyr::select(-c(cases, Grade_Mutational_status)) %>% gather(key = "Gene", value = "GEX", -c(batch, Purity, Index)) %>% dplyr::filter(Gene %in% c("AR", "KLK3", "TMPRSS2", "FOXA1",pca_lncrna))




# res <- do.call(rbind, lapply(unique(df$Gene), function(g){
# 
#   d <- dplyr::filter(df, Gene == g)
#   
#   r <- lm(d$GEX~ d$Purity) %>% residuals()
#   
#   data.frame(batch = d$batch, Index = d$Index, Gene = d$Gene, Corrected_GEX = r) %>% return()
#   
#   
# }))
# df[,'PART1']
# 
# res <- do.call(rbind, lapply(colnames(df)[!colnames(df) %in% c('batch', 'Purity', 'cases', 'Index', 'Grade_Mutational_status')], function(g){
#   print(g)
#   r <- lm(df[,g, drop = T]~ df$Purity) %>% residuals()
#   
#   data.frame(batch = df$batch, Gene = g, Corrected_GEX = r, Cases = df$cases) %>% return()
#   
#   
# }))
# res %>% dplyr::filter(Gene == "AR") %>% group_by(batch) %>% arrange(Corrected_GEX, .by_group = T)
res <-df




res<- mutate(res,Group = ifelse(Gene == "AR", "AR", ifelse(Gene == "SCHLAP1", "SCHLAP1", ifelse(Gene == "PCAT14", "PCAT14", ifelse(Gene == "KLK3", "KLK3", ifelse(Gene == "TMPRSS2", "TMPRSS2", ifelse(Gene == "FOXA1", "FOXA1", "Other PCa lncRNA"))))))) 


res


res%>% ggplot(aes(x = Index, y = GEX, group = Gene, color = Group, label = Gene, alpha = Group)) + geom_line(stat = 'smooth', method = 'loess', se=FALSE, span = 0.25) + scale_color_manual(values = c("darkred", "navy", "darkgreen", 'grey', 'blue', 'orange', 'red')) + geom_vline(xintercept = c(53,389), linetype = 'dotted')  + theme_classic() + scale_alpha_manual(values = c(1,1,1, 0.5, 1, 1, 1), labels =c("AR", "Other PCa lncRNA", "PCAT14", "SCHLAP1", "FOXA1", "KLK3", "TMPRSS2"))

res

dplyr::filter(res, Gene == "AR") %>% ggplot(aes(x  = Index, y = GEX, color = batch)) + geom_point()+ geom_vline(xintercept = c(53,389), linetype = 'dotted')

res
#+ annotate("rect", xmin = 389, xmax = 500, ymin = -4, ymax = 4, alpha = 0.1)

#ggsave(filename = "/Users/ds/Desktop/plot.pdf", width = 10, plot = p)

```

